{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lab 6: Sequence Characteristics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Q0: What's your name?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Andrew Judell-Halfpenny"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The purpose of this exercise is to retrieve your gene's mRNA and protein sequences from Genbank and perform some (more) preliminary analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check that the fasta files containing the 5'UTR, CDS, and 3'UTR sequences of your gene (the files generated in lab exercise 4) are located in your working directory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Identifying the GC content of regions of your mRNA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "from Bio import Entrez # we'll import a set of code written for querying NCBI databases using Entrez\n",
    "from Bio import SeqIO # and a set of code specific for dealing with sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "GI_id = \"NM_000088.3\" # enter the accession # or GeneID for the mRNA sequence of your gene ***as a string***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Documentation for efecth http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch  \n",
    "\n",
    "Code adapted from Biopython Tutorial, 5.3.1  Parsing GenBank records from the net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Entrez.email = \"Andrew.judellHalf001@umb.edu\" # let NCBI know who you are\n",
    "handle = Entrez.efetch(db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=GI_id) # db = database to query, rettype = retrieval type, retmode = retrieval mode \n",
    "nt_seq_record = SeqIO.read(handle, \"gb\") #read the contents \"fetched\" from NCBI\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCGTCGGAGCAGACGGGAGTTTCTCCTCGGGGTCGGAGCAGGAGGCACGCGGAGTGTGAGGCCACGCATGAGCGGACGCTAACCCCCTCCCCAGCCACAAAGAGTCTACATGTCTAGGGTCTAGACATGTTCAGCTTTGTGGACCTCCGGCTCCTGCTCCTCTTAGCGGCCACCGCCCTCCTGACGCACGGCCAAGAGGAAGGCCAAGTCGAGGGCCAAGACGAAGACATCCCACCAATCACCTGCGTACAGAACGGCCTCAGGTACCATGACCGAGACGTGTGGAAACCCGAGCCCTGCCGGATCTGCGTCTGCGACAACGGCAAGGTGTTGTGCGATGACGTGATCTGTGACGAGACCAAGAACTGCCCCGGCGCCGAAGTCCCCGAGGGCGAGTGCTGTCCCGTCTGCCCCGACGGCTCAGAGTCACCCACCGACCAAGAAACCACCGGCGTCGAGGGACCCAAGGGAGACACTGGCCCCCGAGGCCCAAGGGGACCCGCAGGCCCCCCTGGCCGAGATGGCATCCCTGGACAGCCTGGACTTCCCGGACCCCCCGGACCCCCCGGACCTCCCGGACCCCCTGGCCTCGGAGGAAACTTTGCTCCCCAGCTGTCTTATGGCTATGATGAGAAATCAACCGGAGGAATTTCCGTGCCTGGCCCCATGGGTCCCTCTGGTCCTCGTGGTCTCCCTGGCCCCCCTGGTGCACCTGGTCCCCAAGGCTTCCAAGGTCCCCCTGGTGAGCCTGGCGAGCCTGGAGCTTCAGGTCCCATGGGTCCCCGAGGTCCCCCAGGTCCCCCTGGAAAGAATGGAGATGATGGGGAAGCTGGAAAACCTGGTCGTCCTGGTGAGCGTGGGCCTCCTGGGCCTCAGGGTGCTCGAGGATTGCCCGGAACAGCTGGCCTCCCTGGAATGAAGGGACACAGAGGTTTCAGTGGTTTGGATGGTGCCAAGGGAGATGCTGGTCCTGCTGGTCCTAAGGGTGAGCCTGGCAGCCCTGGTGAAAATGGAGCTCCTGGTCAGATGGGCCCCCGTGGCCTGCCTGGTGAGAGAGGTCGCCCTGGAGCCCCTGGCCCTGCTGGTGCTCGTGGAAATGATGGTGCTACTGGTGCTGCCGGGCCCCCTGGTCCCACCGGCCCCGCTGGTCCTCCTGGCTTCCCTGGTGCTGTTGGTGCTAAGGGTGAAGCTGGTCCCCAAGGGCCCCGAGGCTCTGAAGGTCCCCAGGGTGTGCGTGGTGAGCCTGGCCCCCCTGGCCCTGCTGGTGCTGCTGGCCCTGCTGGAAACCCTGGTGCTGATGGACAGCCTGGTGCTAAAGGTGCCAATGGTGCTCCTGGTATTGCTGGTGCTCCTGGCTTCCCTGGTGCCCGAGGCCCCTCTGGACCCCAGGGCCCCGGCGGCCCTCCTGGTCCCAAGGGTAACAGCGGTGAACCTGGTGCTCCTGGCAGCAAAGGAGACACTGGTGCTAAGGGAGAGCCTGGCCCTGTTGGTGTTCAAGGACCCCCTGGCCCTGCTGGAGAGGAAGGAAAGCGAGGAGCTCGAGGTGAACCCGGACCCACTGGCCTGCCCGGACCCCCTGGCGAGCGTGGTGGACCTGGTAGCCGTGGTTTCCCTGGCGCAGATGGTGTTGCTGGTCCCAAGGGTCCCGCTGGTGAACGTGGTTCTCCTGGCCCTGCTGGCCCCAAAGGATCTCCTGGTGAAGCTGGTCGTCCCGGTGAAGCTGGTCTGCCTGGTGCCAAGGGTCTGACTGGAAGCCCTGGCAGCCCTGGTCCTGATGGCAAAACTGGCCCCCCTGGTCCCGCCGGTCAAGATGGTCGCCCCGGACCCCCAGGCCCACCTGGTGCCCGTGGTCAGGCTGGTGTGATGGGATTCCCTGGACCTAAAGGTGCTGCTGGAGAGCCCGGCAAGGCTGGAGAGCGAGGTGTTCCCGGACCCCCTGGCGCTGTCGGTCCTGCTGGCAAAGATGGAGAGGCTGGAGCTCAGGGACCCCCTGGCCCTGCTGGTCCCGCTGGCGAGAGAGGTGAACAAGGCCCTGCTGGCTCCCCCGGATTCCAGGGTCTCCCTGGTCCTGCTGGTCCTCCAGGTGAAGCAGGCAAACCTGGTGAACAGGGTGTTCCTGGAGACCTTGGCGCCCCTGGCCCCTCTGGAGCAAGAGGCGAGAGAGGTTTCCCTGGCGAGCGTGGTGTGCAAGGTCCCCCTGGTCCTGCTGGTCCCCGAGGGGCCAACGGTGCTCCCGGCAACGATGGTGCTAAGGGTGATGCTGGTGCCCCTGGAGCTCCCGGTAGCCAGGGCGCCCCTGGCCTTCAGGGAATGCCTGGTGAACGTGGTGCAGCTGGTCTTCCAGGGCCTAAGGGTGACAGAGGTGATGCTGGTCCCAAAGGTGCTGATGGCTCTCCTGGCAAAGATGGCGTCCGTGGTCTGACTGGCCCCATTGGTCCTCCTGGCCCTGCTGGTGCCCCTGGTGACAAGGGTGAAAGTGGTCCCAGCGGCCCTGCTGGTCCCACTGGAGCTCGTGGTGCCCCCGGAGACCGTGGTGAGCCTGGTCCCCCCGGCCCTGCTGGCTTTGCTGGCCCCCCTGGTGCTGACGGCCAACCTGGTGCTAAAGGCGAACCTGGTGATGCTGGTGCTAAAGGCGATGCTGGTCCCCCTGGCCCTGCCGGACCCGCTGGACCCCCTGGCCCCATTGGTAATGTTGGTGCTCCTGGAGCCAAAGGTGCTCGCGGCAGCGCTGGTCCCCCTGGTGCTACTGGTTTCCCTGGTGCTGCTGGCCGAGTCGGTCCTCCTGGCCCCTCTGGAAATGCTGGACCCCCTGGCCCTCCTGGTCCTGCTGGCAAAGAAGGCGGCAAAGGTCCCCGTGGTGAGACTGGCCCTGCTGGACGTCCTGGTGAAGTTGGTCCCCCTGGTCCCCCTGGCCCTGCTGGCGAGAAAGGATCCCCTGGTGCTGATGGTCCTGCTGGTGCTCCTGGTACTCCCGGGCCTCAAGGTATTGCTGGACAGCGTGGTGTGGTCGGCCTGCCTGGTCAGAGAGGAGAGAGAGGCTTCCCTGGTCTTCCTGGCCCCTCTGGTGAACCTGGCAAACAAGGTCCCTCTGGAGCAAGTGGTGAACGTGGTCCCCCTGGTCCCATGGGCCCCCCTGGATTGGCTGGACCCCCTGGTGAATCTGGACGTGAGGGGGCTCCTGGTGCCGAAGGTTCCCCTGGACGAGACGGTTCTCCTGGCGCCAAGGGTGACCGTGGTGAGACCGGCCCCGCTGGACCCCCTGGTGCTCCTGGTGCTCCTGGTGCCCCTGGCCCCGTTGGCCCTGCTGGCAAGAGTGGTGATCGTGGTGAGACTGGTCCTGCTGGTCCCGCCGGTCCTGTCGGCCCTGTTGGCGCCCGTGGCCCCGCCGGACCCCAAGGCCCCCGTGGTGACAAGGGTGAGACAGGCGAACAGGGCGACAGAGGCATAAAGGGTCACCGTGGCTTCTCTGGCCTCCAGGGTCCCCCTGGCCCTCCTGGCTCTCCTGGTGAACAAGGTCCCTCTGGAGCCTCTGGTCCTGCTGGTCCCCGAGGTCCCCCTGGCTCTGCTGGTGCTCCTGGCAAAGATGGACTCAACGGTCTCCCTGGCCCCATTGGGCCCCCTGGTCCTCGCGGTCGCACTGGTGATGCTGGTCCTGTTGGTCCCCCCGGCCCTCCTGGACCTCCTGGTCCCCCTGGTCCTCCCAGCGCTGGTTTCGACTTCAGCTTCCTGCCCCAGCCACCTCAAGAGAAGGCTCACGATGGTGGCCGCTACTACCGGGCTGATGATGCCAATGTGGTTCGTGACCGTGACCTCGAGGTGGACACCACCCTCAAGAGCCTGAGCCAGCAGATCGAGAACATCCGGAGCCCAGAGGGCAGCCGCAAGAACCCCGCCCGCACCTGCCGTGACCTCAAGATGTGCCACTCTGACTGGAAGAGTGGAGAGTACTGGATTGACCCCAACCAAGGCTGCAACCTGGATGCCATCAAAGTCTTCTGCAACATGGAGACTGGTGAGACCTGCGTGTACCCCACTCAGCCCAGTGTGGCCCAGAAGAACTGGTACATCAGCAAGAACCCCAAGGACAAGAGGCATGTCTGGTTCGGCGAGAGCATGACCGATGGATTCCAGTTCGAGTATGGCGGCCAGGGCTCCGACCCTGCCGATGTGGCCATCCAGCTGACCTTCCTGCGCCTGATGTCCACCGAGGCCTCCCAGAACATCACCTACCACTGCAAGAACAGCGTGGCCTACATGGACCAGCAGACTGGCAACCTCAAGAAGGCCCTGCTCCTCCAGGGCTCCAACGAGATCGAGATCCGCGCCGAGGGCAACAGCCGCTTCACCTACAGCGTCACTGTCGATGGCTGCACGAGTCACACCGGAGCCTGGGGCAAGACAGTGATTGAATACAAAACCACCAAGACCTCCCGCCTGCCCATCATCGATGTGGCCCCCTTGGACGTTGGTGCCCCAGACCAGGAATTCGGCTTCGACGTTGGCCCTGTCTGCTTCCTGTAAACTCCCTCCATCCCAACCTGGCTCCCTCCCACCCAACCAACTTTCCCCCCAACCCGGAAACAGACAAGCAACCCAAACTGAACCCCCTCAAAAGCCAAAAAATGGGAGACAATTTCACATGGACTTTGGAAAATATTTTTTTCCTTTGCATTCATCTCTCAAACTTAGTTTTTATCTTTGACCAACCGAACATGACCAAAAACCAAAAGTGCATTCAACCTTACCAAAAAAAAAAAAAAAAAAAGAATAAATAAATAACTTTTTAAAAAAGGAAGCTTGGTCCACTTGCTTGAAGACCCATGCGGGGGTAAGTCCCTTTCTGCCCGTTGGGCTTATGAAACCCCAATGCTGCCCTTTCTGCTCCTTTCTCCACACCCCCCTTGGGGCCTCCCCTCCACTCCTTCCCAAATCTGTCTCCCCAGAAGACACAGGAAACAATGTATTGTCTGCCCAGCAATCAAAGGCAATGCTCAAACACCCAAGTGGCCCCCACCCTCAGCCCGCTCCTGCCCGCCCAGCACCCCCAGGCCCTGGGGGACCTGGGGTTCTCAGACTGCCAAAGAAGCCTTGCCATCTGGCGCTCCCATGGCTCTTGCAACATCTCCCCTTCGTTTTTGAGGGGGTCATGCCGGGGGAGCCACCAGCCCCTCACTGGGTTCGGAGGAGAGTCAGGAAGGGCCACGACAAAGCAGAAACATCGGATTTGGGGAACGCGTGTCAATCCCTTGTGCCGCAGGGCTGGGCGGGAGAGACTGTTCTGTTCCTTGTGTAACTGTGTTGCTGAAAGACTACCTCGTTCTTGTCTTGATGTGTCACCGGGGCAACTGCCTGGGGGCGGGGATGGGGGCAGGGTGGAAGCGGCTCCCCATTTTATACCAAAGGTGCTACATCTATGTGATGGGTGGGGTGGGGAGGGAATCACTGGTGCTATAGAAATTGAGATGCCCCCCCAGGCCAGCAAATGTTCCTTTTTGTTCAAAGTCTATTTTTATTCCTTGATATTTTTCTTTTTTTTTTTTTTTTTTTGTGGATGGGGACTTGTGAATTTTTCTAAAGGTGCTATTTAACATGGGAGGAGAGCGTGTGCGGCTCCAGCCCAGCCCGCTGCTCACTTTCCACCCTCTCTCCACCTGCCTCTGGCTTCTCAGGCCTCTGCTCTCCGACCTCTCTCCTCTGAAACCCTCCTCCACAGCTGCAGCCCATCCTCCCGGCTCCCTCCTAGTCTGTCCTGCGTCCTCTGTCCCCGGGTTTCAGAGACAACTTCCCAAAGCACAAAGCAGTTTTTCCCCCTAGGGGTGGGAGGAAGCAAAAGACTCTGTACCTATTTTGTATGTGTATAATAATTTGAGATGTTTTTAATTATTTTGATTGCTGGAATAAAGCATGTGGAAATGACCCAAACATAA\n",
      "5927\n"
     ]
    }
   ],
   "source": [
    "ntid = nt_seq_record.id # assign the accession number of the sequence (stored in the id attribute of nt_seq_record) to the variable ntid\n",
    "#print the accession number\n",
    "\n",
    "ntseq = nt_seq_record.seq # assign the sequence (stored in the seq attribute of nt_seq_record) to the variable ntseq \n",
    "print ntseq\n",
    "print len(ntseq)\n",
    "#print the sequence\n",
    "\n",
    "#print the number of nucleotides in the sequence "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we'll count the number of times each nucletide occurs in the sequence. In this lab exercise we will be counting the number of nucleotides over the entire mRNA sequence as well as within the 5'UTR, the CDS, and the 3'UTR. Instead of writing four for loops (one for each sequence/sub-sequence) it is much more efficient to write one function which you can call again and again. \n",
    "\n",
    "Funtions consist of a function definition, \"def\". The entire body of the definintion is indented. \n",
    "\n",
    "See the example function below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input_string: Hello\n",
      "result: Hello World\n"
     ]
    }
   ],
   "source": [
    "def greetings(input_string): # this the defintion of a function called greetings, it takes as an argument a string called input_string\n",
    "    output_string = input_string + \" World\" # the varibable output_string is the sum of input_string and the string \" World\"\n",
    "    return output_string # returns output_string\n",
    "\n",
    "input_string = \"Hello\" # assign input_string the value \"Hello\"\n",
    "result = greetings(input_string) # the variable result is assigned to the result of sending the argument input_string into the function greetings \n",
    "\n",
    "\n",
    "print \"input_string:\", input_string\n",
    "print \"result:\", result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take the following code (written in a previous lab section) and turn it into a function called GC_content which takes a nucleotide sequence seq and returns a list called containing the counts of A's, T's, C's, and G's. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "countA 1064\n",
      "countC 1877\n",
      "countG 1769\n",
      "countT 1217\n",
      "countN 0\n",
      "[1064, 1217, 1877, 1769]\n"
     ]
    }
   ],
   "source": [
    "countA=0 # create variables to hold the count of each nucleotide\n",
    "countC=0\n",
    "countG=0\n",
    "countT=0\n",
    "countN=0 \n",
    "counts = [countA, countT, countC, countG]\n",
    "\n",
    "\n",
    "def ATGC_content(ntseq):\n",
    "# this is the definition of the function GC_content, it input consisting of a nucleotide sequence called sequence\n",
    "   # create a variable to hold the count of any non-standard nucleotides\n",
    "    countA=0 # create variables to hold the count of each nucleotide\n",
    "    countC=0\n",
    "    countG=0\n",
    "    countT=0\n",
    "    countN=0 \n",
    "    counts = [countA, countT, countC, countG]\n",
    "\n",
    "    counts = [countA, countT, countC, countG] \n",
    "    for nt in ntseq:#Loop over each nucleotide in the sequence\n",
    "        if nt == \"C\": # ask if the nt is cytosine \n",
    "            countC +=1 # increase countC by one\n",
    "        elif nt == \"A\": # ask if the nt is adenosine\n",
    "            countA +=1 # increase countA by one\n",
    "        elif nt == \"T\": # ask if the nt is thymine\n",
    "            countT +=1 # increase countT by one\n",
    "        elif nt == \"G\": # ask if the nt is guanine\n",
    "            countG +=1 # increase countG by one\n",
    "        else: # if above not True then nt is not a standard nucleotide\n",
    "            countN +=1 # increase countN by one\n",
    "            print \"Non-standard nt\", nt # print a warning message\n",
    "    counts = [countA, countT, countC, countG]\n",
    "    print \"countA\", countA # print the values of the count variables\n",
    "    print \"countC\", countC\n",
    "    print \"countG\", countG\n",
    "    print \"countT\", countT\n",
    "    print \"countN\", countN\n",
    "     # generate a list containing the number of A, T, C, and G in the sequence.\n",
    "    print counts # return the list counts\n",
    "    return counts\n",
    "    \n",
    "counts = ATGC_content(ntseq)\n",
    "[countA, countT, countC, countG] = counts\n",
    "# assign the value of the variable ntseq to the variable sequence\n",
    "# the variable counts is assigned to the result of sending the argument sequence into the function GC_content"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Q1: Explain the idea behind \"reusable code\". Why is this so important? Name at least two means by which you can make your code more reusable."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Reusable code saves time by alleviating the need to rewrite the same code multiple times.  Code can be made more reusable by using understandable syntax and simple code."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, calculate the number of cytosine and guanine nucleotides as a percentage of the total number of nt's in the sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1064, 1217, 1877, 1769]\n",
      "1064\n",
      "0.615151003881\n"
     ]
    }
   ],
   "source": [
    "print counts\n",
    "print countA\n",
    "countGC = countG +countC\n",
    "countall= float(countA + countC + countG + countT + countN)\n",
    "GCperc = countGC / float(countall)\n",
    "print GCperc\n",
    " \n",
    "# create a variable countGC to hold the count of G's and C's in the sequence\n",
    "# define a variable, percGC, which is the countGC/float(total number of nt)*100\n",
    "# print the value of percGC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Q2: What does assert do and why is it useful?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "assert is a failsafe.  It can be used at the begining of a code to prevent an infinite loop or for debugging."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate a list containing the number of A, T, C, G as a proportion of the total number of nucleotides."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "percent G 29.8464653282\n",
      "percent C 31.6686350599\n",
      "percent A 17.951746246\n",
      "percent T 20.533153366\n"
     ]
    }
   ],
   "source": [
    "for x in counts:\n",
    "    Gper = countG/float(countall)*100\n",
    "    Cper = countC/float(countall)*100\n",
    "    Tper = countT/float(countall)*100\n",
    "    Aper = countA/float(countall)*100\n",
    "propor = {float(Cper), float(Gper), float(Tper), float(Aper)}\n",
    "assert float(Cper + Gper + Tper + Aper)==100\n",
    "print \"percent G\", Gper\n",
    "print \"percent C\", Cper\n",
    "print \"percent A\", Aper\n",
    "print \"percent T\", Tper\n",
    "\n",
    "# generate a list called propor to hold the proportions\n",
    "# for each element x in counts\n",
    "    # define a variable, p, which is the count/float(total number of nt)\n",
    "    # add a the calculated proportion to the list propor\n",
    "\n",
    "    \n",
    "# print the list propor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can create a graph - with the four nucleotides on the x-axis and the percentages on the y-axis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXwAAAEKCAYAAAARnO4WAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAG+RJREFUeJzt3Xm4XXV97/H3JwnzEEKABASCiAKFMvgIj7cgHEIRrBUo\nQphaQYaL9qLArTKINpFSB6xpRRxaGZqgAhZldIDaZFfAWwYZDLMgBBByEBJIGERIvveP328nKzvn\nZO+Ts/ZZ52R9Xs+zn7P2Gr/rd9b+7t/6rrX3VkRgZmarv1FVB2BmZkPDCd/MrCac8M3MasIJ38ys\nJpzwzcxqwgnfzKwmnPBHIElPSJrchfVuJWmhJPUzfaqky8ve7nDWrk3yPIskbTN0UYGktSXdIOkl\nSVcN5bZt5HLC7xJJT0rqlbROYdyJkmZXGVdR6xtHRDwdERvGyj+cUasPbrS2iaTZkk5omWeDiHhy\niEM7HNgUGBcRR7ZOlDRN0hJJhxfGjc7jts7P/z0//1DLsv+cx3+kXRCSjsnH+iJJP5K0UWHampIu\nlfSypGclndGy7G6S7pL0qqQ7Je3aMv0MSc/lN7WLJa1RmDZJ0o8lzc/r/rqkUYXpUyQ9mLd9v6RD\nVrIPlxe287CkE9vt90jlhN89QWrf0/sYbzZYk4BHV/LmHMCLwOdbzk6iZfgRYGlilzQaOAJ4rF0A\nknYCvg0cC0wAXge+VZjl88A7gK2AycCZkt6fl10DuBaYCWyU/14naUyefiBwJrBf3td35PU1fRN4\nPm93N2Bf4G/zslsAlwOnR8TYvJ7vS9qkn135IvD2iNgIOBg4X9Lu7fZ/JHLC766vAH8nacPWCbmH\nsqSlV7Jc71HSybmXsjD3UnbrYz2SdLakxyT9XtKVLb2sg/Oy8yXNkrR9Hj8T2Bq4Ia//U60xSdpG\nUiP3km4CNmnZ9nsl3SZpgaR7JO3bX0NI2lLSDyU9n+O8sBD/Z3MvcV7udW7Y0kbHS3pK0ouSTpH0\nHkn35X36emEbx0m6Nff2XsptN7kwfXNJ1+X1PCrppMK0PXIv8+Xc2/un1v+TpPOB9wEX5TZr7sMS\nSdvm4Q0lzcz7+YSkc1viu0XSV3Lsj0s6aCVttkM+JhZImqPcE5c0Dfh74Kgcx0f7WcVNwB+Bvymu\ntmWeG4G9JY3Nzw8C7gPm9RdXwTHA9RFxW0S8BnwOOEzSenn6R4DzImJhRDwM/BtwfJ62HzA6Ii6M\niDcj4us5tsmFZS+JiIcj4mXgPKC4n9sAV+Vlnwd+BuyUp20JLIiImwEi4ifAq6Q3jRVExIMR8YdC\n+0R/8454EeFHFx7AE6SD92rgH/K4E4FZeXgSsBgYVVhmNnBCHj4CeBp4d36+LbBVcd15+DTgl8Dm\nwBqkHtb387R3Aa/kOEYDnwZ+A4wprGe/wvaXiymv9yt5ve8DFgIz87S3AS8AB+bn++fn4/toi1HA\nvcA/AWsDawJ/lqedADyat70u8MPCNiYBS0i9uTWBPyf1In8EjAe2AHqB9+X5jwPeBD6Z93cK8BKw\nUZ7+C+DreX92JfUQewr7emweXhfYs582Wfo/KuzfYmDbPDwTuCavYxKpB/3RQnxv5H0W8DHgd/0c\nP2Py/+qsPLxfbv935ulTm+3Uz/JTcyx/CTye22N0bs+t8zyXkRLpt4FT8rirgCOBW4CPtDnGrwU+\n3TJuIbA7qde+BNi0MO0w4L48fDrw45ZlrwfOyMP3AkcUpm2c23lcfn4y8O/AOqRjcQ5wcOF4m533\nfRRwKPAUsM5K9uUbpDeFJcBdwLpV55BuPNzD776pwKmSxg9wuROBCyLiboCI+G1EPN3HfKcA50bE\ncxHxJukFfHjupU8BboyIWRGxmJRw1wH+rLB8fxdotwbeA/x9pF7ULcANhVmOJb1gb8rx/RfphfIX\nfaxuT9Ib0pkR8YeI+GNE/DJPOwaYHhFzI/USzyH1XJvHZpB6iX+MiJ+TXpRXRMSLEfEsKTEVT797\nI/UaF0fED0gJ94OStgT+F3BW3p/7gItZVs54E9hO0viIeC0i7uirXfqh3GajSMny7LyOucBXWb6H\nPTciLo2UZWYAEyVt1sc63wusFxFfjoi3ImI2qTd+9ADiIiJuBH4PnLSS2S4Hjsu9/H1IibwT6wMv\nt4xbCGyQp0XL9Oa0dsv2NX0hqZ2b028Bds7jnwLujIjrASJiSd6nK0hvsN8lvaG93t+ORMT/ydvc\nm9SheKO/eUcyJ/wui4gHSC/Ucwa46Faknlk7k4BrcolgPvAgKXlNIPWA5xZiCdJZw9s6WO/mpNPi\n4otkbmF4EjCluV1JC4C98nJ97cvc/EJstVyMeXhMjr/p+cLw66ReffH5+oXnv2tZ/9y8jS2A+flN\npTit2RYnANsDD0u6XdIH+4i1nU1y7E/1sw0olEpy26ol/qYtSP+rotZ1deqzwLmks6sVRMRtpAvA\n55I6CJ0mu1eA1nLlWGBRnkbL9Oa0dsv2NX0s6Q1kkSSRSjhXk86kNgE2lvRlAEl/DlwA7BMRawA9\nwCWSdlnZzkTyS9Lx+vGVzTtSOeEPjWmkU9Dii/XV/HfdwriJheGn6ayO+BTwgYjYOD/GRcR6EfEc\n8CwpMRdtBTyTh1d2Afk5YJwKdxmRav7F+Ga2bHeDiLigj3U9DWxdvF5R0BrjJNIbVm8f83aiNSFu\nnbfxLCkprNcy7XcAEfF4RBwTEZuSksXVLfvetLI2eyHH3ro/rW9CnXiW9L8qWhrvQOQzo8dIFzX7\ni/+7wP8lnXV06gFSaQwASe8glcsejYiXSMdQ8c6bXfMyzWVbE/AuwP19rZt0YbY3IhaQyjtbAd/I\nZ2sLSOWpDxS2898RcQ9ARNwF3E4qCXZiDKtpDd8JfwhExOOk2ugnC+NeIL14/zpfEDyB5Q+yi4FP\nSXo3pBeTpNYEAPCvwBe07Fa7TSUdnKf9gFTO2E/SGEmfAv4A/L88fR7p2kCRcnxPkUo0n5e0hqS9\ngeLte98FPiTp/Tn+tSXtq3SHRKs7SC/+L0laV9JakpplpSuAM/IF4vWBfwSuLJwN9Hv/ez82k/SJ\nvL9HADuQSk/PkOr0X8zb34VUNrs8t9uxWnYXx8ukxNhXDL2s2GbA0lLCD4B/lLS+pEnAGc1tDNDt\nwGuSzsz70kOqSV+xCuuC1Ms/cyXTLwQOiIhbB7DO75GOgb3yG+l5wA8jotmZuRz4rKSNJO1I6vRc\nlqc1gMX5f7WmpE+S2rt52/JM4ERJO0oal+O/DCAiXiRdf/qY0q2mG5Guj9yXl72TdCF6VwClO272\nBn7dugP59XKkpPXycXwgcBTw8wG0w8hR9UWE1fUB/JZ8YTU/3xJ4DfivwrgD83zzSRdHl7sgCPxv\n4GFSnfLXwK6t6yYlo9PzfC+TLvSdX1jHIaTe0oK8/h0L0w4mlQnmk3p3rRcotyFd6FxIuuPjQgoX\nCoE9SC/cF0mJ8AZgy37aY0vSxcwXSCWafynE/1nSmUovqYc5Nk/r68L2U6RT9ebzmcBn8vBxpNru\nhaSLtQ8D+xfm3SLH+GJup5ML0y7P219IugD4ob5iINXWH8nraO5D8aLtRnldz+e2PbewjeOAX7S0\ny9Jl+2izHXP7vkTq+R5cmNbRRduWcT8G3qLlom0/y/+CNhdt83xH5f1cRKp9b1SYtiZwCem4fA44\nrWXZXUmdilfz311app9O6pS8ROoArVGYtgvpeJ6f2/pKlr9A/Lf5f/wy6ezm9MK0c8gXjEnloEZe\nz0ukN40T2u33SH0o73RXSFqLdOCsSTpNujoiPi9pKundvlmb/UxE/KxrgVgtSDoOODEi9qk6FrPh\naEw3Vx4Rb0jaLyJeU/pAx22SfponT4+I6d3cvpmZLdP1Gn4suytiLdIbTPOUYqC1WTMbYpK+pfS1\nCQvzozn8zapjs4HrakkHlt6b/CvSBclvRMQ5uaRzPKm+dhfwd5E+TWdmZl3S9YS/dEPp4/LXAJ8g\nfRDkhYiI/HH1zSNitf3CIjOz4WDIEj6ApM8BrxZr9/nWtRsiYoUPRUgauuDMzFYjEbFC2byrNXxJ\nmzS/lCl/iOUA0icZix8wOoxlH7ZYQdW3MXXymDp1auUxrE4Pt6fbcrg+Rkp79qerd+mQPmY/I9fx\nR5G+3e4n+dsEdyN90OJJ0vfBmJlZF3X7tsw5wLv7GN/2hxXMzKxc/mqFEvT09FQdwmrF7Vket2W5\nRnp7DulF24GSFMM5PjOz4UgSMdQXbc3MbPhwwjczqwknfDOzmnDCNzOrCSd8M7OacMI3M6sJJ3wz\ns5pwwjczqwknfDOzmnDCNzOrCSd8M7OacMI3M6sJJ3wzs5pwwjczqwknfDOzmuj2TxzaMDJx4jb0\n9s6tOoxhYcKEScyb92TVYZgNKf8ASo1IAtyeiVb6Y89mI5l/AMXMrOac8M3MasIJ38ysJpzwzcxq\noqsJX9Jakm6XdI+kOZKm5vHjJN0s6RFJN0ka2804zMxsCO7SkbRuRLwmaTRwG/BJ4MPAixFxgaSz\ngHERcXYfy/ounRL5Lp0i36Vjq6/K7tKJiNfy4Fqk+/4DOASYkcfPAA7tdhxmZnXX9YQvaZSke4B5\nwH9GxJ3AhIjoBYiIecBm3Y7DzKzuuv5J24hYAuwuaUPgGkk7sWJdod9z62nTpi0d7unpoaenpwtR\nmpmNXI1Gg0aj0Xa+If2kraTPAa8BJwE9EdEraSIwOyJ27GN+1/BL5Bp+kWv4w4m/9mOZMr72o5Ia\nvqRNmnfgSFoHOAB4CLgeOD7PdhxwXTfjMLPhLSX78IPo6htft0s6mwMzJI0ivblcFRE/kfQ/wA8k\nnQDMBaZ0OQ4zs9rzl6fViEs6RS7pDCc+NosGf2z6y9PMzGrOCd/MrCac8M3MasIJ38ysJpzwzcxq\nwgnfzKwmnPDNzGrCCd/MrCac8M3MasIJ38ysJpzwzcxqwgnfzKwmnPDNzGrCCd/MrCac8M3MasIJ\n38ysJpzwzcxqwgnfzKwmnPDNzGrCCd/MrCac8M3MasIJ38ysJpzwzcxqoqsJX9KWkmZJekDSHEmf\nyOOnSnpG0t35cVA34zAzM1BEdG/l0kRgYkTcK2l94FfAIcCRwKKImN5m+ehmfHUjCXB7JsLH1vDh\nY7No8MemJCJCrePHDGqtbUTEPGBeHn5F0kPA25oxdXPbZma2vCGr4UvaBtgNuD2POlXSvZIuljR2\nqOIwM6urrvbwm3I552rgtNzT/yZwXkSEpPOB6cCJfS07bdq0pcM9PT309PR0P2AzsxGk0WjQaDTa\nztfVGj6ApDHAjcBPI+JrfUyfBNwQEbv0Mc01/BK5TlrkGv5w4mOzqHs1/KEo6VwKPFhM9vlibtNh\nwP1DEIeZWa11+y6dvYBfAHNIb98BfAY4hlTPXwI8CZwSEb19LO8efonciypyD3848bFZ1L0eftdL\nOoPhhF8uv6iKnPCHEx+bRSO7pGNmZsOAE76ZWU044ZuZ1YQTvplZTTjhm5nVhBO+mVlNOOGbmdWE\nE76ZWU044ZuZ1YQTvplZTTjhm5nVhBO+mVlNOOGbmdWEE76ZWU044ZuZ1YQTvplZTTjhm5nVhBO+\nmVlNOOGbmdWEE76ZWU044ZuZ1cSYTmaS9C7g08Ck4jIRMblLcZmZWck6SvjAfwDfBr4DLO5eOGZm\n1i2dJvy3IuJbA125pC2BmcAEYAnwnYi4UNI44CrSGcOTwJSIeHmg6zczs84pItrPJE0DngeuAd5o\njo+I+W2WmwhMjIh7Ja0P/Ao4BPgo8GJEXCDpLGBcRJzdx/LRSXzWGUmA2zMRgzm2Jk7cht7euSXG\nM3JNmDCJefOeHNQ6fGwWDe7YhNSeEaEVxneY8J/oY3RExLYDDOJa4KL82DcievObQiMiduhjfif8\nEvlFVTS4F5XbsqicBOX2bKo44ZdB0jZAA9gZeDoixhWmzY+IjftYxgm/RH5RFTnhl8cJv1zdS/id\n3qWzBvBxYJ88qgH8a0S82eHy6wNXA6dFxCuSWvem372bNm3a0uGenh56eno62aSZWW00Gg0ajUbb\n+Tot6VwMrAHMyKP+BlgcESd1sOwY4EbgpxHxtTzuIaCnUNKZHRE79rGse/glci+qyD388riHX66K\ne/jAHhGxa+H5LEn3dbjspcCDzWSfXQ8cD3wZOA64rsN1mZnZKur0k7aLJb2j+UTStnRwP76kvYBj\ngcmS7pF0t6SDSIn+AEmPAPsDXxp46GZmNhCdlnT2By4DfguIdP/8RyNidleDc0mnVD5tLnJJpzwu\n6ZRrGNylI2ktYPv89JGIeGNl85fBCb9cflEVOeGXxwm/XBXV8CVNjohZkg5rmbRdXuGPBhWVmZkN\nmXYXbfcFZgEf6mNaAE74ZmYjRKc1/LdHxBPtxpXNJZ1y+bS5yCWd8rikU67ulXQ6vUvnh32Mu3pQ\nEZmZ2ZBqV8PfAdgJGNtSx98QWLubgZmZWbna1fC3B/4S2Ijl6/iLgJO7FZSZmZWvbQ1f0mjgrIj4\nwtCEtNy2XcMvkeukRa7hl8c1/HJV//XId0TEnoOKYBX08SVrteXvHC+bE355nPDLVX3C/2fSl6dd\nBbzaHB8Rdw8qqvbbDR8ETX5RlcsJvzw+NstVfcLv6ysUots/Yu6EX+QXVbmc8MvjY7Ncw+CrFarg\nhF/kF1W5nPDL42OzXBXfhy9prKTpku7Kj69KGjuoiMzMbEh1+sGrS0m3Yk7Jj4Wkb880M7MRotMa\n/r0RsVu7cWVzSafIp83lckmnPD42y1X9Vyu8Lmnvwsr2Al4fVERmZjakOv2Jw48DM3LdXsB80k8T\nmpnZCDGgu3QkbQgQEQu7FtHy23NJZymfNpfLJZ3y+NgsV/V36YyXdCHQAGZL+pqk8YOKyMzMhlSn\nNfwrgd8DHwYOz8NXdSsoMzMrX6d36dwfETu3jJsTEX/atchwSWd5Pm0ul0s65fGxWa7q79K5WdJR\nkkblxxTgpkFFZGZmQ6rTHv4iYD1gSR41imVfohYRsWFXgnMPv8C9qHK5h18eH5vlqriHHxEbRMSo\niBiTH6PyuA1WluwlXSKpV9KvC+OmSnpG0t35cdCq7ZKZmQ1Ex7dlSjoY2Cc/bUTEjR0sszfwCjAz\nInbJ46YCiyJiegfLu4e/lHtR5XIPvzw+NstV/W2ZXwJOAx7Mj9MkfbHdchFxK7Cgr1V2sl0zMytP\npxdt/wI4ICIujYhLgYOADw5iu6dKulfSxf7WTTOzodHpVytA+iHz+Xl4MEn6m8B5ERGSzgemAyf2\nP/u0wnBPfpiZWVOj0aDRaLSdr9O7dI4GvgTMJpVj9gHOjoi2H76SNAm4oVnD73Ranu4a/lKuk5bL\nNfzy+NgsV/dq+G17+Er/iVuB9wJ75NFnRcS8TrdNoWYvaWJh2cOA+ztcj5mZDUKnPfxV+lStpO+T\najDjgV5gKrAfsBvpnv4ngVMioref5d3DX8q9qHK5h18eH5vlqv5HzGcAF0XEnYOKYoCc8Iv8oiqX\nE355fGyWq/qE/zDwTlKP/FVSiSb6q72XxQm/yC+qcjnhl8fHZrkqrOFnBw5q62ZmVrmVJnxJawMf\nA7YD5gCXRMRbQxGYmZmVq90Hr2YA7yEl+w8AX+16RGZm1hXtSjp/0rw7R9IlwB3dD8nMzLqhXQ//\nzeaASzlmZiPbSu/SkbSYZd97L2Ad4DWW3aXTle/BL2zfd+ks5TshyuW7dMrjY7NcFd2lExGjB7VV\nMzMbNjr9tkwzMxvhnPDNzGrCCd/MrCac8M3MasIJ38ysJpzwzcxqwgnfzKwmnPDNzGrCCd/MrCac\n8M3MasIJ38ysJpzwzcxqwgnfzKwmnPDNzGrCCd/MrCa6mvAlXSKpV9KvC+PGSbpZ0iOSbpI0tpsx\nmJlZ0u0e/mXAgS3jzgZ+HhHbA7OAc7ocg5mZ0eWEHxG3AgtaRh8CzMjDM4BDuxmDmZklVdTwN4uI\nXoCImAdsVkEMZma1s9LftB0ibX6td1phuCc/zMysqdFo0Gg02s6nwf46etsNSJOAGyJil/z8IaAn\nInolTQRmR8SO/Swb/iX7pnJ+yd7t2TS49nRbFvnYLFc57RkRah0/FCUd5UfT9cDxefg44LohiMHM\nrPa62sOX9H1SDWY80AtMBa4F/gPYCpgLTImIl/pZ3j38pdyLKpd7+OXxsVmu7vXwu17SGQwn/CK/\nqMrlhF8eH5vlGtklHTMzGwac8M3MasIJ38ysJpzwzcxqwgnfzKwmnPDNzGrCCd/MrCac8M3MasIJ\n38ysJpzwzcxqwgnfzKwmnPDNzGrCCd/MrCac8M3MasIJ38ysJpzwzcxqwgnfzKwmnPDNzGrCCd/M\nrCac8M3MasIJ38ysJpzwzcxqwgnfzKwmxlS1YUlPAi8DS4A3I2LPqmIxM6uDyhI+KdH3RMSCCmMw\nM6uNKks6qnj7Zma1UmXCDeA/Jd0p6eQK4zAzq4UqSzp7RcRzkjYlJf6HIuLWFWebVhjuyQ8zM2tq\nNBo0Go228ykiuh9NuyCkqcCiiJjeMj7SiYCBGOz/ShJuz6bBtafbssjHZrnKac+IUOv4Sko6ktaV\ntH4eXg94P3B/FbGYmdVFVSWdCcA1qQfPGOB7EXFzRbGYmdXCsCjp9MclnSKfNpfLJZ3y+Ngs12pW\n0jEzs6HnhG9mVhNO+GZmNeGEb2ZWE074ZmY14YRvZlYTTvhmZjXhhG9mVhNO+GZmNeGEb2ZWE074\nZmY14YRvZlYTTvhmZjXhhG9mVhNO+GZmNeGEb2ZWE074ZmY14YRvZlYTTvhmZjXhhG9mVhNO+GZm\nNeGEb2ZWE5UlfEkHSXpY0qOSzqoqDjOzuqgk4UsaBVwEHAjsBBwtaYcqYilHo+oAVjONqgNYjTSq\nDmA106g6gEGpqoe/J/CbiJgbEW8CVwKHVBRLCRpVB7CaaVQdwGqkUXUAq5lG1QEMSlUJ/23A04Xn\nz+RxZmbWJb5oa2ZWE4qIod+o9F5gWkQclJ+fDUREfLllvqEPzsxsNRARah1XVcIfDTwC7A88B9wB\nHB0RDw15MGZmNTGmio1GxGJJpwI3k8pKlzjZm5l1VyU9fDMzG3q+aDsIkg6VtETSu6qOZSSTtLGk\neyTdLek5Sc8UnldyFjrSSZog6QpJv5F0p6QbJW1XdVwjlaTNJH1P0mO5PW+TNOJuJXfCH5yjgFuA\no6sOZCSLiPkRsXtEvBv4FjC9+Twi3qo6vhHqGmBWRLwzIvYAzgEmVBzTSHYt0IiI7XJ7HgVsWXFM\nA+aEv4okrQfsBZyIE36ZVrizwAZG0n7AHyPiO81xETEnIm6rMKwRS9Jk4I2W9nw6Ir5RYVirxAl/\n1R0C/CwiHgNekLR71QGZZTsDv6o6iNXITsDdVQdRBif8VXc06SshAK4CjqkwFjMbIpIuknSvpNur\njmWgfEFsFUgaB0wGds4fDhsNBPDpSgMzSx4ADq86iNXIA8CHm08i4lRJ44E7qwtp1biHv2qOAGZG\nxNsjYtuImAQ8IWnvqgMzi4hZwJqSTmqOk/SnkvaqMKwRK7fnWpJOKYxer6p4BsMJf9UcSboLouhH\n+OKtDR9/BRyQbyOcA3wBmFdxTCPZoUCPpMcl/Q9wGXBmxTENmD94ZWZWE+7hm5nVhBO+mVlNOOGb\nmdWEE76ZWU044ZuZ1YQTvplZTTjhm5nVhBO+mVlN/H/mPXWFJ3k9rAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f0b9c4e5a90>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt # import a set of code written for generating graphics and tell the computer how to display graphs\n",
    "%pylab inline \n",
    "\n",
    "nucleotides = arange(4) # the x-axis has four positions, one for each of the four nucleotides\n",
    "bar(nucleotides, propor) # chart will be a bar chart, containing nucleotides on the x-axis and count as percent of length of the sequence on the y-axis\n",
    "xticks(nucleotides + 0.5,  ('A', 'T', 'C', 'G') ) # label the x-axis with the nucleotides in the same order as in the counts list\n",
    "title(\"Nucleotide composition of \" + ntid) #add a title to the graph\n",
    "ylabel(\"Proportion\") #label the y-axis\n",
    "show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next test if the nucleotide composition of your gene differs significantly from the expected equal representation of nucleotides using the chi-square test. Remember that the chi-square test statistic is the sum over all values (here each of the four nucleotides) of the difference between the number of times that nucleotide occurs within your sequence (\"the observed\") and the number of times you would expect to see that nucleotide if the nucleotides occur with equal frequencies (\"the expected\"), squared, divided by the expected. The smaller the difference between the observed and the expected, the smaller the test statistic will be and the less support you have for rejecting the null hypothesis that the observed and expected data come from the same distribution, i.e. are equally distributed. Assuming that the null hypothesis is true, the p-value represents the probability of finding a test statistic equal to or greater than your test statistic by chance. A p-value of 0.05 or less is the generally accepted standard for significance in scientific publications. Where the p-value is less than 0.05 we reject the null hypothesis; if the p-value is greater than 0.05 we fail to reject the null hypothesis. An excellent video reviewing the concept can be found here - http://youtu.be/HwD7ekD5l0g (ignore the part using the statistical software \"R\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expected number of nucleotides:  [1482, 1482, 1482, 1482]\n",
      "Observed number of nucleotides:  [1064, 1217, 1877, 1769]\n",
      "Chi-squared test statistic:  326.142375169\n",
      "p-value:  2.18299273447e-70\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats # import a set of code written for computing statistical tests\n",
    "\n",
    "def chi_calcs(sequence, counts): # define a function named chi_calcs for performing a chi-squared test on the nucleotide frequency within a sequence\n",
    "    expfreq = [int(round(0.25*len(sequence),0)), int(round(0.25*len(sequence),0)), int(round(0.25*len(sequence),0)), int(round(0.25*len(sequence),0))] # the expected number of each nucleotide in a sequence the length of your sequence at equal frequency \n",
    "    print \"Expected number of nucleotides: \", expfreq # print out the list containing the expected frequency of each nucleotide\n",
    "    obsfreq = counts # assign the observed frequency of each nucleotide in the list counts to the variable named obsfreq\n",
    "    print \"Observed number of nucleotides: \", obsfreq # print out the list containing the observed frequency of each nucleotide\n",
    "    assert len(expfreq) == len(obsfreq) # assert that the number of items in the two lists are the equal\n",
    "\n",
    "    chi2, pval = stats.chisquare(obsfreq, expfreq) # calculates the chi2 test statistic (chi2) and the p-value (pval) using the observed and expected frequency lists\n",
    "    print \"Chi-squared test statistic: \", chi2 # print the value of the chi-square test statistic\n",
    "    print \"p-value: \", pval # print the p-value\n",
    "chi_calcs(ntseq, counts)\n",
    "# assign the value of the variable ntseq to the variable sequence\n",
    "# send the arguments sequence and observed frequencies of nucleotides to the function chi_calcs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Q3: Considering your mRNA sequence as a whole, are the four nucleotides present in equal frequencies?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "No, the p-value is below the alpha thresh hold of reasonably explainable by the null hypothesis and therefor not present in equal frequencies."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "source": [
    "As we've seen, the nucleotide content of a gene can differ significantly over the length of an mRNA sequence. Using the fasta files you generated last week, calculate the nucleotide content over the 5'UTR, CDS, and 3'UTR regions of your gene.\n",
    "\n",
    "We'll use Biopython's SeqIO module to parse the fasta files into their header's and sequence's.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NM_000088.3_5UTR\n",
      "TCGTCGGAGCAGACGGGAGTTTCTCCTCGGGGTCGGAGCAGGAGGCACGCGGAGTGTGAGGCCACGCATGAGCGGACGCTAACCCCCTCCCCAGCCACAAAGAGTCTACATGTCTAGGGTCTAGAC\n",
      "126\n",
      "\n",
      "\n",
      "NM_000088.3_CDS\n",
      "ATGTTCAGCTTTGTGGACCTCCGGCTCCTGCTCCTCTTAGCGGCCACCGCCCTCCTGACGCACGGCCAAGAGGAAGGCCAAGTCGAGGGCCAAGACGAAGACATCCCACCAATCACCTGCGTACAGAACGGCCTCAGGTACCATGACCGAGACGTGTGGAAACCCGAGCCCTGCCGGATCTGCGTCTGCGACAACGGCAAGGTGTTGTGCGATGACGTGATCTGTGACGAGACCAAGAACTGCCCCGGCGCCGAAGTCCCCGAGGGCGAGTGCTGTCCCGTCTGCCCCGACGGCTCAGAGTCACCCACCGACCAAGAAACCACCGGCGTCGAGGGACCCAAGGGAGACACTGGCCCCCGAGGCCCAAGGGGACCCGCAGGCCCCCCTGGCCGAGATGGCATCCCTGGACAGCCTGGACTTCCCGGACCCCCCGGACCCCCCGGACCTCCCGGACCCCCTGGCCTCGGAGGAAACTTTGCTCCCCAGCTGTCTTATGGCTATGATGAGAAATCAACCGGAGGAATTTCCGTGCCTGGCCCCATGGGTCCCTCTGGTCCTCGTGGTCTCCCTGGCCCCCCTGGTGCACCTGGTCCCCAAGGCTTCCAAGGTCCCCCTGGTGAGCCTGGCGAGCCTGGAGCTTCAGGTCCCATGGGTCCCCGAGGTCCCCCAGGTCCCCCTGGAAAGAATGGAGATGATGGGGAAGCTGGAAAACCTGGTCGTCCTGGTGAGCGTGGGCCTCCTGGGCCTCAGGGTGCTCGAGGATTGCCCGGAACAGCTGGCCTCCCTGGAATGAAGGGACACAGAGGTTTCAGTGGTTTGGATGGTGCCAAGGGAGATGCTGGTCCTGCTGGTCCTAAGGGTGAGCCTGGCAGCCCTGGTGAAAATGGAGCTCCTGGTCAGATGGGCCCCCGTGGCCTGCCTGGTGAGAGAGGTCGCCCTGGAGCCCCTGGCCCTGCTGGTGCTCGTGGAAATGATGGTGCTACTGGTGCTGCCGGGCCCCCTGGTCCCACCGGCCCCGCTGGTCCTCCTGGCTTCCCTGGTGCTGTTGGTGCTAAGGGTGAAGCTGGTCCCCAAGGGCCCCGAGGCTCTGAAGGTCCCCAGGGTGTGCGTGGTGAGCCTGGCCCCCCTGGCCCTGCTGGTGCTGCTGGCCCTGCTGGAAACCCTGGTGCTGATGGACAGCCTGGTGCTAAAGGTGCCAATGGTGCTCCTGGTATTGCTGGTGCTCCTGGCTTCCCTGGTGCCCGAGGCCCCTCTGGACCCCAGGGCCCCGGCGGCCCTCCTGGTCCCAAGGGTAACAGCGGTGAACCTGGTGCTCCTGGCAGCAAAGGAGACACTGGTGCTAAGGGAGAGCCTGGCCCTGTTGGTGTTCAAGGACCCCCTGGCCCTGCTGGAGAGGAAGGAAAGCGAGGAGCTCGAGGTGAACCCGGACCCACTGGCCTGCCCGGACCCCCTGGCGAGCGTGGTGGACCTGGTAGCCGTGGTTTCCCTGGCGCAGATGGTGTTGCTGGTCCCAAGGGTCCCGCTGGTGAACGTGGTTCTCCTGGCCCTGCTGGCCCCAAAGGATCTCCTGGTGAAGCTGGTCGTCCCGGTGAAGCTGGTCTGCCTGGTGCCAAGGGTCTGACTGGAAGCCCTGGCAGCCCTGGTCCTGATGGCAAAACTGGCCCCCCTGGTCCCGCCGGTCAAGATGGTCGCCCCGGACCCCCAGGCCCACCTGGTGCCCGTGGTCAGGCTGGTGTGATGGGATTCCCTGGACCTAAAGGTGCTGCTGGAGAGCCCGGCAAGGCTGGAGAGCGAGGTGTTCCCGGACCCCCTGGCGCTGTCGGTCCTGCTGGCAAAGATGGAGAGGCTGGAGCTCAGGGACCCCCTGGCCCTGCTGGTCCCGCTGGCGAGAGAGGTGAACAAGGCCCTGCTGGCTCCCCCGGATTCCAGGGTCTCCCTGGTCCTGCTGGTCCTCCAGGTGAAGCAGGCAAACCTGGTGAACAGGGTGTTCCTGGAGACCTTGGCGCCCCTGGCCCCTCTGGAGCAAGAGGCGAGAGAGGTTTCCCTGGCGAGCGTGGTGTGCAAGGTCCCCCTGGTCCTGCTGGTCCCCGAGGGGCCAACGGTGCTCCCGGCAACGATGGTGCTAAGGGTGATGCTGGTGCCCCTGGAGCTCCCGGTAGCCAGGGCGCCCCTGGCCTTCAGGGAATGCCTGGTGAACGTGGTGCAGCTGGTCTTCCAGGGCCTAAGGGTGACAGAGGTGATGCTGGTCCCAAAGGTGCTGATGGCTCTCCTGGCAAAGATGGCGTCCGTGGTCTGACTGGCCCCATTGGTCCTCCTGGCCCTGCTGGTGCCCCTGGTGACAAGGGTGAAAGTGGTCCCAGCGGCCCTGCTGGTCCCACTGGAGCTCGTGGTGCCCCCGGAGACCGTGGTGAGCCTGGTCCCCCCGGCCCTGCTGGCTTTGCTGGCCCCCCTGGTGCTGACGGCCAACCTGGTGCTAAAGGCGAACCTGGTGATGCTGGTGCTAAAGGCGATGCTGGTCCCCCTGGCCCTGCCGGACCCGCTGGACCCCCTGGCCCCATTGGTAATGTTGGTGCTCCTGGAGCCAAAGGTGCTCGCGGCAGCGCTGGTCCCCCTGGTGCTACTGGTTTCCCTGGTGCTGCTGGCCGAGTCGGTCCTCCTGGCCCCTCTGGAAATGCTGGACCCCCTGGCCCTCCTGGTCCTGCTGGCAAAGAAGGCGGCAAAGGTCCCCGTGGTGAGACTGGCCCTGCTGGACGTCCTGGTGAAGTTGGTCCCCCTGGTCCCCCTGGCCCTGCTGGCGAGAAAGGATCCCCTGGTGCTGATGGTCCTGCTGGTGCTCCTGGTACTCCCGGGCCTCAAGGTATTGCTGGACAGCGTGGTGTGGTCGGCCTGCCTGGTCAGAGAGGAGAGAGAGGCTTCCCTGGTCTTCCTGGCCCCTCTGGTGAACCTGGCAAACAAGGTCCCTCTGGAGCAAGTGGTGAACGTGGTCCCCCTGGTCCCATGGGCCCCCCTGGATTGGCTGGACCCCCTGGTGAATCTGGACGTGAGGGGGCTCCTGGTGCCGAAGGTTCCCCTGGACGAGACGGTTCTCCTGGCGCCAAGGGTGACCGTGGTGAGACCGGCCCCGCTGGACCCCCTGGTGCTCCTGGTGCTCCTGGTGCCCCTGGCCCCGTTGGCCCTGCTGGCAAGAGTGGTGATCGTGGTGAGACTGGTCCTGCTGGTCCCGCCGGTCCTGTCGGCCCTGTTGGCGCCCGTGGCCCCGCCGGACCCCAAGGCCCCCGTGGTGACAAGGGTGAGACAGGCGAACAGGGCGACAGAGGCATAAAGGGTCACCGTGGCTTCTCTGGCCTCCAGGGTCCCCCTGGCCCTCCTGGCTCTCCTGGTGAACAAGGTCCCTCTGGAGCCTCTGGTCCTGCTGGTCCCCGAGGTCCCCCTGGCTCTGCTGGTGCTCCTGGCAAAGATGGACTCAACGGTCTCCCTGGCCCCATTGGGCCCCCTGGTCCTCGCGGTCGCACTGGTGATGCTGGTCCTGTTGGTCCCCCCGGCCCTCCTGGACCTCCTGGTCCCCCTGGTCCTCCCAGCGCTGGTTTCGACTTCAGCTTCCTGCCCCAGCCACCTCAAGAGAAGGCTCACGATGGTGGCCGCTACTACCGGGCTGATGATGCCAATGTGGTTCGTGACCGTGACCTCGAGGTGGACACCACCCTCAAGAGCCTGAGCCAGCAGATCGAGAACATCCGGAGCCCAGAGGGCAGCCGCAAGAACCCCGCCCGCACCTGCCGTGACCTCAAGATGTGCCACTCTGACTGGAAGAGTGGAGAGTACTGGATTGACCCCAACCAAGGCTGCAACCTGGATGCCATCAAAGTCTTCTGCAACATGGAGACTGGTGAGACCTGCGTGTACCCCACTCAGCCCAGTGTGGCCCAGAAGAACTGGTACATCAGCAAGAACCCCAAGGACAAGAGGCATGTCTGGTTCGGCGAGAGCATGACCGATGGATTCCAGTTCGAGTATGGCGGCCAGGGCTCCGACCCTGCCGATGTGGCCATCCAGCTGACCTTCCTGCGCCTGATGTCCACCGAGGCCTCCCAGAACATCACCTACCACTGCAAGAACAGCGTGGCCTACATGGACCAGCAGACTGGCAACCTCAAGAAGGCCCTGCTCCTCCAGGGCTCCAACGAGATCGAGATCCGCGCCGAGGGCAACAGCCGCTTCACCTACAGCGTCACTGTCGATGGCTGCACGAGTCACACCGGAGCCTGGGGCAAGACAGTGATTGAATACAAAACCACCAAGACCTCCCGCCTGCCCATCATCGATGTGGCCCCCTTGGACGTTGGTGCCCCAGACCAGGAATTCGGCTTCGACGTTGGCCCTGTCTGCTTCCTGTAA\n",
      "4395\n",
      "\n",
      "\n",
      "NM_000088.3_3UTR\n",
      "ACTCCCTCCATCCCAACCTGGCTCCCTCCCACCCAACCAACTTTCCCCCCAACCCGGAAACAGACAAGCAACCCAAACTGAACCCCCTCAAAAGCCAAAAAATGGGAGACAATTTCACATGGACTTTGGAAAATATTTTTTTCCTTTGCATTCATCTCTCAAACTTAGTTTTTATCTTTGACCAACCGAACATGACCAAAAACCAAAAGTGCATTCAACCTTACCAAAAAAAAAAAAAAAAAAAGAATAAATAAATAACTTTTTAAAAAAGGAAGCTTGGTCCACTTGCTTGAAGACCCATGCGGGGGTAAGTCCCTTTCTGCCCGTTGGGCTTATGAAACCCCAATGCTGCCCTTTCTGCTCCTTTCTCCACACCCCCCTTGGGGCCTCCCCTCCACTCCTTCCCAAATCTGTCTCCCCAGAAGACACAGGAAACAATGTATTGTCTGCCCAGCAATCAAAGGCAATGCTCAAACACCCAAGTGGCCCCCACCCTCAGCCCGCTCCTGCCCGCCCAGCACCCCCAGGCCCTGGGGGACCTGGGGTTCTCAGACTGCCAAAGAAGCCTTGCCATCTGGCGCTCCCATGGCTCTTGCAACATCTCCCCTTCGTTTTTGAGGGGGTCATGCCGGGGGAGCCACCAGCCCCTCACTGGGTTCGGAGGAGAGTCAGGAAGGGCCACGACAAAGCAGAAACATCGGATTTGGGGAACGCGTGTCAATCCCTTGTGCCGCAGGGCTGGGCGGGAGAGACTGTTCTGTTCCTTGTGTAACTGTGTTGCTGAAAGACTACCTCGTTCTTGTCTTGATGTGTCACCGGGGCAACTGCCTGGGGGCGGGGATGGGGGCAGGGTGGAAGCGGCTCCCCATTTTATACCAAAGGTGCTACATCTATGTGATGGGTGGGGTGGGGAGGGAATCACTGGTGCTATAGAAATTGAGATGCCCCCCCAGGCCAGCAAATGTTCCTTTTTGTTCAAAGTCTATTTTTATTCCTTGATATTTTTCTTTTTTTTTTTTTTTTTTTGTGGATGGGGACTTGTGAATTTTTCTAAAGGTGCTATTTAACATGGGAGGAGAGCGTGTGCGGCTCCAGCCCAGCCCGCTGCTCACTTTCCACCCTCTCTCCACCTGCCTCTGGCTTCTCAGGCCTCTGCTCTCCGACCTCTCTCCTCTGAAACCCTCCTCCACAGCTGCAGCCCATCCTCCCGGCTCCCTCCTAGTCTGTCCTGCGTCCTCTGTCCCCGGGTTTCAGAGACAACTTCCCAAAGCACAAAGCAGTTTTTCCCCCTAGGGGTGGGAGGAAGCAAAAGACTCTGTACCTATTTTGTATGTGTATAATAATTTGAGATGTTTTTAATTATTTTGATTGCTGGAATAAAGCATGTGGAAATGACCCAAACATAA\n",
      "1406\n",
      "\n",
      "\n",
      "['NM_000088.3_3UTR', 'NM_000088.3_CDS', 'NM_000088.3_5UTR']\n",
      "[Seq('ACTCCCTCCATCCCAACCTGGCTCCCTCCCACCCAACCAACTTTCCCCCCAACC...TAA', SingleLetterAlphabet()), Seq('ATGTTCAGCTTTGTGGACCTCCGGCTCCTGCTCCTCTTAGCGGCCACCGCCCTC...TAA', SingleLetterAlphabet()), Seq('TCGTCGGAGCAGACGGGAGTTTCTCCTCGGGGTCGGAGCAGGAGGCACGCGGAG...GAC', SingleLetterAlphabet())]\n",
      "1406\n",
      "4395\n",
      "126\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "five_UTR_filename = ntid + \"_5UTR_seq.fasta\" # the file where you saved your gene's 5'UTR sequence\n",
    "CDS_filename = ntid + \"_CDS_seq.fasta\" # the file where you saved your gene's CDS\n",
    "three_UTR_filename = ntid + \"_3UTR_seq.fasta\" # the file where you saved your gene's 3'UTR sequence\n",
    "\n",
    "gene_regions = {} # create a dictionary called gene_regions\n",
    "fasta_list = [five_UTR_filename, CDS_filename, three_UTR_filename] # put the file names into a list\n",
    "\n",
    "for fl in fasta_list: # for each file, fl, in the list, fasta_list\n",
    "    for s in SeqIO.parse(fl, \"fasta\"): # for each fasta record, s, in the file, fl, parse the fasta records (separates the text into the fasta header and sequence)\n",
    "        print s.id # print the fasta header of the fasta record\n",
    "        print s.seq # print the sequence contained in the fasta record\n",
    "        print len(s.seq)\n",
    "        gene_regions[s.id] = s.seq\n",
    "        print \"\\n\"\n",
    "print gene_regions.keys()\n",
    "print gene_regions.values()\n",
    "for j in gene_regions.values():\n",
    "    print len(j)\n",
    "        \n",
    "        # convert the sequence to a string\n",
    "        # add an element to the dictionary, gene_regions, where the key is the fasta header and the value is the string of the sequence\n",
    "        # print the length of the fasta sequence\n",
    "        # print a newline to create a visual separation between the sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each sequence in the dictionary gene_regions, call the GC_content function. Remember that GC_content returns a list of the counts of each nucleotide in the sequence. Call the chi_calcs function, giving it the sequence and the list of counts - this will print out the observed and expected number of nucleotides, the chi-squared test statistic, and corresponding p-value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['NM_000088.3_3UTR', 'NM_000088.3_CDS', 'NM_000088.3_5UTR'])\n",
      "dict_values([Seq('ACTCCCTCCATCCCAACCTGGCTCCCTCCCACCCAACCAACTTTCCCCCCAACC...TAA', SingleLetterAlphabet()), Seq('ATGTTCAGCTTTGTGGACCTCCGGCTCCTGCTCCTCTTAGCGGCCACCGCCCTC...TAA', SingleLetterAlphabet()), Seq('TCGTCGGAGCAGACGGGAGTTTCTCCTCGGGGTCGGAGCAGGAGGCACGCGGAG...GAC', SingleLetterAlphabet())])\n",
      "\n",
      "\n",
      "countA 334\n",
      "countC 404\n",
      "countG 304\n",
      "countT 364\n",
      "countN 0\n",
      "[334, 364, 404, 304]\n",
      "Expected number of nucleotides:  [352, 352, 352, 352]\n",
      "Observed number of nucleotides:  [334, 364, 404, 304]\n",
      "Chi-squared test statistic:  15.5568181818\n",
      "p-value:  0.00139765242955\n",
      "\n",
      "\n",
      "countA 703\n",
      "countC 1436\n",
      "countG 1423\n",
      "countT 833\n",
      "countN 0\n",
      "[703, 833, 1436, 1423]\n",
      "Expected number of nucleotides:  [1099, 1099, 1099, 1099]\n",
      "Observed number of nucleotides:  [703, 833, 1436, 1423]\n",
      "Chi-squared test statistic:  405.929936306\n",
      "p-value:  1.1499153285e-87\n",
      "\n",
      "\n",
      "countA 27\n",
      "countC 37\n",
      "countG 42\n",
      "countT 20\n",
      "countN 0\n",
      "[27, 20, 37, 42]\n",
      "Expected number of nucleotides:  [32, 32, 32, 32]\n",
      "Observed number of nucleotides:  [27, 20, 37, 42]\n",
      "Chi-squared test statistic:  9.1875\n",
      "p-value:  0.0268991008988\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "keys = gene_regions.viewkeys()\n",
    "value = gene_regions.viewvalues()\n",
    "print keys\n",
    "print value\n",
    "print \"\\n\"\n",
    "for j in value:\n",
    "    countA=0 \n",
    "    countC=0\n",
    "    countG=0\n",
    "    countT=0\n",
    "    countN=0 \n",
    "    for k in j:\n",
    "        if k == \"C\": \n",
    "            countC +=1 \n",
    "        elif k == \"A\": # ask if the nt is adenosine\n",
    "            countA +=1 # increase countA by one\n",
    "        elif k == \"T\": # ask if the nt is thymine\n",
    "            countT +=1 # increase countT by one\n",
    "        elif k == \"G\": # ask if the nt is guanine\n",
    "            countG +=1 # increase countG by one\n",
    "        else: # if above not True then nt is not a standard nucleotide\n",
    "            countN +=1 # increase countN by one\n",
    "            print \"Non-standard nt\", nt # print a warning message\n",
    "    counts = [countA, countT, countC, countG]\n",
    "    print \"countA\", countA # print the values of the count variables\n",
    "    print \"countC\", countC\n",
    "    print \"countG\", countG\n",
    "    print \"countT\", countT\n",
    "    print \"countN\", countN\n",
    "    # generate a list containing the number of A, T, C, and G in the sequence.\n",
    "    print counts # return the list counts\n",
    "    chi_calcs(j,counts)\n",
    "    print \"\\n\"\n",
    "    \n",
    "# for each element of the dictionary gene_regions\n",
    "    # print the dictionary key, the sequence id\n",
    "    # assign the nucleotide seq to a variable called sequence\n",
    "    # print a message with the number of nucleotides in the sequence\n",
    "    #counts = GC_content(sequence) # send the sequence to the function GC_content, returns a list called counts containing the counts of each nucleotide  \n",
    "    #chi_calcs(sequence, counts) # send the sequence and list counts to the function chi_calcs, this will print out the expected and observed nucletides, the chi-square statistic and the p-value\n",
    "    # print a newline as a visual separation between records\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Q4: Can you reject the null hypothesis of equal distribution of nucleotides for any of the regions of your gene? Explain. If you reject the null hypothesis, which nucletides appear to be over or under represented?"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "For each the 3', 5', and the CDS, the null hypothesis can be rejected.  The 3' sequence has an over representation of C and T to a lesser extent , the 5' sequence has an overrepresentation of G and C to a lesser extent, and the CDS has an overrepresentation of CG (unsurprisingly)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "source": [
    "Now, let's generate a graph showing the percent GC of 50 nt bins accross your gene's entire mRNA sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5927\n",
      "TCGTCGGAGCAGACGGGAGTTTCTCCTCGGGGTCGGAGCAGGAGGCACGCGGAGTGTGAGGCCACGCATGAGCGGACGCTAACCCCCTCCCCAGCCACAAAGAGTCTACATGTCTAGGGTCTAGACATGTTCAGCTTTGTGGACCTCCGGCTCCTGCTCCTCTTAGCGGCCACCGCCCTCCTGACGCACGGCCAAGAGGAAGGCCAAGTCGAGGGCCAAGACGAAGACATCCCACCAATCACCTGCGTACAGAACGGCCTCAGGTACCATGACCGAGACGTGTGGAAACCCGAGCCCTGCCGGATCTGCGTCTGCGACAACGGCAAGGTGTTGTGCGATGACGTGATCTGTGACGAGACCAAGAACTGCCCCGGCGCCGAAGTCCCCGAGGGCGAGTGCTGTCCCGTCTGCCCCGACGGCTCAGAGTCACCCACCGACCAAGAAACCACCGGCGTCGAGGGACCCAAGGGAGACACTGGCCCCCGAGGCCCAAGGGGACCCGCAGGCCCCCCTGGCCGAGATGGCATCCCTGGACAGCCTGGACTTCCCGGACCCCCCGGACCCCCCGGACCTCCCGGACCCCCTGGCCTCGGAGGAAACTTTGCTCCCCAGCTGTCTTATGGCTATGATGAGAAATCAACCGGAGGAATTTCCGTGCCTGGCCCCATGGGTCCCTCTGGTCCTCGTGGTCTCCCTGGCCCCCCTGGTGCACCTGGTCCCCAAGGCTTCCAAGGTCCCCCTGGTGAGCCTGGCGAGCCTGGAGCTTCAGGTCCCATGGGTCCCCGAGGTCCCCCAGGTCCCCCTGGAAAGAATGGAGATGATGGGGAAGCTGGAAAACCTGGTCGTCCTGGTGAGCGTGGGCCTCCTGGGCCTCAGGGTGCTCGAGGATTGCCCGGAACAGCTGGCCTCCCTGGAATGAAGGGACACAGAGGTTTCAGTGGTTTGGATGGTGCCAAGGGAGATGCTGGTCCTGCTGGTCCTAAGGGTGAGCCTGGCAGCCCTGGTGAAAATGGAGCTCCTGGTCAGATGGGCCCCCGTGGCCTGCCTGGTGAGAGAGGTCGCCCTGGAGCCCCTGGCCCTGCTGGTGCTCGTGGAAATGATGGTGCTACTGGTGCTGCCGGGCCCCCTGGTCCCACCGGCCCCGCTGGTCCTCCTGGCTTCCCTGGTGCTGTTGGTGCTAAGGGTGAAGCTGGTCCCCAAGGGCCCCGAGGCTCTGAAGGTCCCCAGGGTGTGCGTGGTGAGCCTGGCCCCCCTGGCCCTGCTGGTGCTGCTGGCCCTGCTGGAAACCCTGGTGCTGATGGACAGCCTGGTGCTAAAGGTGCCAATGGTGCTCCTGGTATTGCTGGTGCTCCTGGCTTCCCTGGTGCCCGAGGCCCCTCTGGACCCCAGGGCCCCGGCGGCCCTCCTGGTCCCAAGGGTAACAGCGGTGAACCTGGTGCTCCTGGCAGCAAAGGAGACACTGGTGCTAAGGGAGAGCCTGGCCCTGTTGGTGTTCAAGGACCCCCTGGCCCTGCTGGAGAGGAAGGAAAGCGAGGAGCTCGAGGTGAACCCGGACCCACTGGCCTGCCCGGACCCCCTGGCGAGCGTGGTGGACCTGGTAGCCGTGGTTTCCCTGGCGCAGATGGTGTTGCTGGTCCCAAGGGTCCCGCTGGTGAACGTGGTTCTCCTGGCCCTGCTGGCCCCAAAGGATCTCCTGGTGAAGCTGGTCGTCCCGGTGAAGCTGGTCTGCCTGGTGCCAAGGGTCTGACTGGAAGCCCTGGCAGCCCTGGTCCTGATGGCAAAACTGGCCCCCCTGGTCCCGCCGGTCAAGATGGTCGCCCCGGACCCCCAGGCCCACCTGGTGCCCGTGGTCAGGCTGGTGTGATGGGATTCCCTGGACCTAAAGGTGCTGCTGGAGAGCCCGGCAAGGCTGGAGAGCGAGGTGTTCCCGGACCCCCTGGCGCTGTCGGTCCTGCTGGCAAAGATGGAGAGGCTGGAGCTCAGGGACCCCCTGGCCCTGCTGGTCCCGCTGGCGAGAGAGGTGAACAAGGCCCTGCTGGCTCCCCCGGATTCCAGGGTCTCCCTGGTCCTGCTGGTCCTCCAGGTGAAGCAGGCAAACCTGGTGAACAGGGTGTTCCTGGAGACCTTGGCGCCCCTGGCCCCTCTGGAGCAAGAGGCGAGAGAGGTTTCCCTGGCGAGCGTGGTGTGCAAGGTCCCCCTGGTCCTGCTGGTCCCCGAGGGGCCAACGGTGCTCCCGGCAACGATGGTGCTAAGGGTGATGCTGGTGCCCCTGGAGCTCCCGGTAGCCAGGGCGCCCCTGGCCTTCAGGGAATGCCTGGTGAACGTGGTGCAGCTGGTCTTCCAGGGCCTAAGGGTGACAGAGGTGATGCTGGTCCCAAAGGTGCTGATGGCTCTCCTGGCAAAGATGGCGTCCGTGGTCTGACTGGCCCCATTGGTCCTCCTGGCCCTGCTGGTGCCCCTGGTGACAAGGGTGAAAGTGGTCCCAGCGGCCCTGCTGGTCCCACTGGAGCTCGTGGTGCCCCCGGAGACCGTGGTGAGCCTGGTCCCCCCGGCCCTGCTGGCTTTGCTGGCCCCCCTGGTGCTGACGGCCAACCTGGTGCTAAAGGCGAACCTGGTGATGCTGGTGCTAAAGGCGATGCTGGTCCCCCTGGCCCTGCCGGACCCGCTGGACCCCCTGGCCCCATTGGTAATGTTGGTGCTCCTGGAGCCAAAGGTGCTCGCGGCAGCGCTGGTCCCCCTGGTGCTACTGGTTTCCCTGGTGCTGCTGGCCGAGTCGGTCCTCCTGGCCCCTCTGGAAATGCTGGACCCCCTGGCCCTCCTGGTCCTGCTGGCAAAGAAGGCGGCAAAGGTCCCCGTGGTGAGACTGGCCCTGCTGGACGTCCTGGTGAAGTTGGTCCCCCTGGTCCCCCTGGCCCTGCTGGCGAGAAAGGATCCCCTGGTGCTGATGGTCCTGCTGGTGCTCCTGGTACTCCCGGGCCTCAAGGTATTGCTGGACAGCGTGGTGTGGTCGGCCTGCCTGGTCAGAGAGGAGAGAGAGGCTTCCCTGGTCTTCCTGGCCCCTCTGGTGAACCTGGCAAACAAGGTCCCTCTGGAGCAAGTGGTGAACGTGGTCCCCCTGGTCCCATGGGCCCCCCTGGATTGGCTGGACCCCCTGGTGAATCTGGACGTGAGGGGGCTCCTGGTGCCGAAGGTTCCCCTGGACGAGACGGTTCTCCTGGCGCCAAGGGTGACCGTGGTGAGACCGGCCCCGCTGGACCCCCTGGTGCTCCTGGTGCTCCTGGTGCCCCTGGCCCCGTTGGCCCTGCTGGCAAGAGTGGTGATCGTGGTGAGACTGGTCCTGCTGGTCCCGCCGGTCCTGTCGGCCCTGTTGGCGCCCGTGGCCCCGCCGGACCCCAAGGCCCCCGTGGTGACAAGGGTGAGACAGGCGAACAGGGCGACAGAGGCATAAAGGGTCACCGTGGCTTCTCTGGCCTCCAGGGTCCCCCTGGCCCTCCTGGCTCTCCTGGTGAACAAGGTCCCTCTGGAGCCTCTGGTCCTGCTGGTCCCCGAGGTCCCCCTGGCTCTGCTGGTGCTCCTGGCAAAGATGGACTCAACGGTCTCCCTGGCCCCATTGGGCCCCCTGGTCCTCGCGGTCGCACTGGTGATGCTGGTCCTGTTGGTCCCCCCGGCCCTCCTGGACCTCCTGGTCCCCCTGGTCCTCCCAGCGCTGGTTTCGACTTCAGCTTCCTGCCCCAGCCACCTCAAGAGAAGGCTCACGATGGTGGCCGCTACTACCGGGCTGATGATGCCAATGTGGTTCGTGACCGTGACCTCGAGGTGGACACCACCCTCAAGAGCCTGAGCCAGCAGATCGAGAACATCCGGAGCCCAGAGGGCAGCCGCAAGAACCCCGCCCGCACCTGCCGTGACCTCAAGATGTGCCACTCTGACTGGAAGAGTGGAGAGTACTGGATTGACCCCAACCAAGGCTGCAACCTGGATGCCATCAAAGTCTTCTGCAACATGGAGACTGGTGAGACCTGCGTGTACCCCACTCAGCCCAGTGTGGCCCAGAAGAACTGGTACATCAGCAAGAACCCCAAGGACAAGAGGCATGTCTGGTTCGGCGAGAGCATGACCGATGGATTCCAGTTCGAGTATGGCGGCCAGGGCTCCGACCCTGCCGATGTGGCCATCCAGCTGACCTTCCTGCGCCTGATGTCCACCGAGGCCTCCCAGAACATCACCTACCACTGCAAGAACAGCGTGGCCTACATGGACCAGCAGACTGGCAACCTCAAGAAGGCCCTGCTCCTCCAGGGCTCCAACGAGATCGAGATCCGCGCCGAGGGCAACAGCCGCTTCACCTACAGCGTCACTGTCGATGGCTGCACGAGTCACACCGGAGCCTGGGGCAAGACAGTGATTGAATACAAAACCACCAAGACCTCCCGCCTGCCCATCATCGATGTGGCCCCCTTGGACGTTGGTGCCCCAGACCAGGAATTCGGCTTCGACGTTGGCCCTGTCTGCTTCCTGTAAACTCCCTCCATCCCAACCTGGCTCCCTCCCACCCAACCAACTTTCCCCCCAACCCGGAAACAGACAAGCAACCCAAACTGAACCCCCTCAAAAGCCAAAAAATGGGAGACAATTTCACATGGACTTTGGAAAATATTTTTTTCCTTTGCATTCATCTCTCAAACTTAGTTTTTATCTTTGACCAACCGAACATGACCAAAAACCAAAAGTGCATTCAACCTTACCAAAAAAAAAAAAAAAAAAAGAATAAATAAATAACTTTTTAAAAAAGGAAGCTTGGTCCACTTGCTTGAAGACCCATGCGGGGGTAAGTCCCTTTCTGCCCGTTGGGCTTATGAAACCCCAATGCTGCCCTTTCTGCTCCTTTCTCCACACCCCCCTTGGGGCCTCCCCTCCACTCCTTCCCAAATCTGTCTCCCCAGAAGACACAGGAAACAATGTATTGTCTGCCCAGCAATCAAAGGCAATGCTCAAACACCCAAGTGGCCCCCACCCTCAGCCCGCTCCTGCCCGCCCAGCACCCCCAGGCCCTGGGGGACCTGGGGTTCTCAGACTGCCAAAGAAGCCTTGCCATCTGGCGCTCCCATGGCTCTTGCAACATCTCCCCTTCGTTTTTGAGGGGGTCATGCCGGGGGAGCCACCAGCCCCTCACTGGGTTCGGAGGAGAGTCAGGAAGGGCCACGACAAAGCAGAAACATCGGATTTGGGGAACGCGTGTCAATCCCTTGTGCCGCAGGGCTGGGCGGGAGAGACTGTTCTGTTCCTTGTGTAACTGTGTTGCTGAAAGACTACCTCGTTCTTGTCTTGATGTGTCACCGGGGCAACTGCCTGGGGGCGGGGATGGGGGCAGGGTGGAAGCGGCTCCCCATTTTATACCAAAGGTGCTACATCTATGTGATGGGTGGGGTGGGGAGGGAATCACTGGTGCTATAGAAATTGAGATGCCCCCCCAGGCCAGCAAATGTTCCTTTTTGTTCAAAGTCTATTTTTATTCCTTGATATTTTTCTTTTTTTTTTTTTTTTTTTGTGGATGGGGACTTGTGAATTTTTCTAAAGGTGCTATTTAACATGGGAGGAGAGCGTGTGCGGCTCCAGCCCAGCCCGCTGCTCACTTTCCACCCTCTCTCCACCTGCCTCTGGCTTCTCAGGCCTCTGCTCTCCGACCTCTCTCCTCTGAAACCCTCCTCCACAGCTGCAGCCCATCCTCCCGGCTCCCTCCTAGTCTGTCCTGCGTCCTCTGTCCCCGGGTTTCAGAGACAACTTCCCAAAGCACAAAGCAGTTTTTCCCCCTAGGGGTGGGAGGAAGCAAAAGACTCTGTACCTATTTTGTATGTGTATAATAATTTGAGATGTTTTTAATTATTTTGATTGCTGGAATAAAGCATGTGGAAATGACCCAAACATAA\n"
     ]
    }
   ],
   "source": [
    "[x0, x1, x2] = value\n",
    "print len(ntseq)\n",
    "print ntseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{1: 68.0, 2: 66.0, 3: 50.0, 4: 68.0, 5: 57.99999999999999, 6: 62.0, 7: 57.99999999999999, 8: 68.0, 9: 66.0, 10: 74.0, 11: 72.0, 12: 78.0, 13: 46.0, 14: 70.0, 15: 68.0, 16: 72.0, 17: 54.0, 18: 68.0, 19: 56.00000000000001, 20: 64.0, 21: 64.0, 22: 66.0, 23: 76.0, 24: 60.0, 25: 74.0, 26: 68.0, 27: 56.00000000000001, 28: 80.0, 29: 64.0, 30: 56.00000000000001, 31: 66.0, 32: 74.0, 33: 66.0, 34: 62.0, 35: 64.0, 36: 68.0, 37: 76.0, 38: 57.99999999999999, 39: 72.0, 40: 68.0, 41: 68.0, 42: 62.0, 43: 68.0, 44: 64.0, 45: 70.0, 46: 72.0, 47: 60.0, 48: 60.0, 49: 64.0, 50: 66.0, 51: 74.0, 52: 72.0, 53: 57.99999999999999, 54: 68.0, 55: 70.0, 56: 64.0, 57: 66.0, 58: 68.0, 59: 66.0, 60: 62.0, 61: 64.0, 62: 60.0, 63: 70.0, 64: 66.0, 65: 70.0, 66: 74.0, 67: 64.0, 68: 80.0, 69: 60.0, 70: 72.0, 71: 64.0, 72: 62.0, 73: 72.0, 74: 72.0, 75: 60.0, 76: 60.0, 77: 62.0, 78: 70.0, 79: 54.0, 80: 52.0, 81: 62.0, 82: 50.0, 83: 60.0, 84: 66.0, 85: 60.0, 86: 60.0, 87: 64.0, 88: 64.0, 89: 50.0, 90: 62.0, 91: 62.0, 92: 54.0, 93: 44.0, 94: 22.0, 95: 40.0, 96: 14.000000000000002, 97: 57.99999999999999, 98: 56.00000000000001, 99: 60.0, 100: 44.0, 101: 78.0, 102: 62.0, 103: 56.00000000000001, 104: 70.0, 105: 50.0, 106: 57.99999999999999, 107: 54.0, 108: 64.0, 109: 56.00000000000001, 110: 44.0, 111: 16.0, 112: 40.0, 113: 68.0, 114: 60.0, 115: 68.0, 116: 54.0, 117: 48.0, 118: 20.0, 119: 20.0}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#break into size 50 chunks\n",
    "region = {}\n",
    "for i in range(0,len(ntseq)/50):\n",
    "    subset = ntseq[i*50:(i+1)*50]\n",
    "    region[i] = subset \n",
    "region[i+1] = ntseq[(i+1)*50:]\n",
    "regionGC = {}\n",
    "Y = 0\n",
    "gcnum = 0\n",
    "for i in region:\n",
    "    Y += 1\n",
    "    gcnum = 0\n",
    "    for x in ntseq[(i*50):(i+1)*50]:\n",
    "        if x in \"G\":\n",
    "            gcnum += 1\n",
    "        if x in \"C\":\n",
    "            gcnum += 1\n",
    "    gcpercent = (float(gcnum)/50)*100\n",
    "    regionGC[Y] = gcpercent\n",
    "print regionGC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5927\n",
      "5927\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'int' object has no attribute 'seq'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-138-476b9bd39276>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;32mprint\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m: 'int' object has no attribute 'seq'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'd like to plot the GC content by region over the length with the coding region of your gene highlighted. To do this we first need to know the begining and ending position of the coding region. Write a for loop using the features attribute of nt_seq_record; if the feature type is \"CDS\", assign the starting position to cds_start and the ending position to cds_end."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NM_000088.3\n",
      "Name: NM_000088\n",
      "Description: Homo sapiens collagen, type I, alpha 1 (COL1A1), mRNA.\n",
      "Number of features: 89\n",
      "/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "reference sequence was derived from AC015909.14 and Z74615.1.\n",
      "This sequence is a reference standard in the RefSeqGene project.\n",
      "On Jul 14, 2006 this sequence version replaced gi:14719826.\n",
      "Summary: This gene encodes the pro-alpha1 chains of type I collagen\n",
      "whose triple helix comprises two alpha1 chains and one alpha2\n",
      "chain. Type I is a fibril-forming collagen found in most connective\n",
      "tissues and is abundant in bone, cornea, dermis and tendon.\n",
      "Mutations in this gene are associated with osteogenesis imperfecta\n",
      "types I-IV, Ehlers-Danlos syndrome type VIIA, Ehlers-Danlos\n",
      "syndrome Classical type, Caffey Disease and idiopathic\n",
      "osteoporosis. Reciprocal translocations between chromosomes 17 and\n",
      "22, where this gene and the gene for platelet-derived growth factor\n",
      "beta are located, are associated with a particular type of skin\n",
      "tumor called dermatofibrosarcoma protuberans, resulting from\n",
      "unregulated expression of the growth factor. Two transcripts,\n",
      "resulting from the use of alternate polyadenylation signals, have\n",
      "been identified for this gene. [provided by R. Dalgleish, Feb\n",
      "2008].\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "##Evidence-Data-START##\n",
      "Transcript exon combination :: Z74615.1, BC036531.2 [ECO:0000332]\n",
      "RNAseq introns              :: single sample supports all introns\n",
      "                               SAMEA1968189, SAMEA1968540\n",
      "                               [ECO:0000348]\n",
      "##Evidence-Data-END##\n",
      "COMPLETENESS: full length.\n",
      "/sequence_version=3\n",
      "/source=Homo sapiens (human)\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/keywords=['RefSeq']\n",
      "/references=[Reference(title='Synergistic effect of pro-inflammatory TNFalpha and IL-17 in periostin mediated collagen deposition: potential role in liver fibrosis', ...), Reference(title='MiR-10a and miR-181c regulate collagen type I generation in hypertrophic scars by targeting PAI-1 and uPA', ...), Reference(title='[Mutation analysis and prenatal diagnosis of COL1A1 gene in a Chinese family with type I osteogenesis imperfecta]', ...), Reference(title='[Identification of a novel splicing mutation in COL1A1 gene in a Chinese family affected with typeosteogenesis imperfecta]', ...), Reference(title='Caffey Disease', ...), Reference(title='COL1A1/2-Related Osteogenesis Imperfecta', ...), Reference(title='Activation of expression of genes coding for extracellular matrix proteins in Tat-producing glioblastoma cells', ...), Reference(title='Short amino acid sequences derived from C1q receptor (C1q-R) show homology with the alpha chains of fibronectin and vitronectin receptors and collagen type IV', ...), Reference(title='Evidence for a relationship between Ehlers-Danlos type VII C in humans and bovine dermatosparaxis', ...), Reference(title='Defective folding and stable association with protein disulfide isomerase/prolyl hydroxylase of type I procollagen with a deletion in the pro alpha 2(I) chain that preserves the Gly-X-Y repeat pattern', ...)]\n",
      "/accessions=['NM_000088']\n",
      "/data_file_division=PRI\n",
      "/date=21-MAY-2015\n",
      "/organism=Homo sapiens\n",
      "/gi=110349771\n",
      "Seq('TCGTCGGAGCAGACGGGAGTTTCTCCTCGGGGTCGGAGCAGGAGGCACGCGGAG...TAA', IUPACAmbiguousDNA())\n",
      "<class 'Bio.SeqRecord.SeqRecord'>\n",
      "126\n",
      "4521\n",
      "ATGTTCAGCTTTGTGGACCTCCGGCTCCTGCTCCTCTTAGCGGCCACCGCCCTCCTGACGCACGGCCAAGAGGAAGGCCAAGTCGAGGGCCAAGACGAAGACATCCCACCAATCACCTGCGTACAGAACGGCCTCAGGTACCATGACCGAGACGTGTGGAAACCCGAGCCCTGCCGGATCTGCGTCTGCGACAACGGCAAGGTGTTGTGCGATGACGTGATCTGTGACGAGACCAAGAACTGCCCCGGCGCCGAAGTCCCCGAGGGCGAGTGCTGTCCCGTCTGCCCCGACGGCTCAGAGTCACCCACCGACCAAGAAACCACCGGCGTCGAGGGACCCAAGGGAGACACTGGCCCCCGAGGCCCAAGGGGACCCGCAGGCCCCCCTGGCCGAGATGGCATCCCTGGACAGCCTGGACTTCCCGGACCCCCCGGACCCCCCGGACCTCCCGGACCCCCTGGCCTCGGAGGAAACTTTGCTCCCCAGCTGTCTTATGGCTATGATGAGAAATCAACCGGAGGAATTTCCGTGCCTGGCCCCATGGGTCCCTCTGGTCCTCGTGGTCTCCCTGGCCCCCCTGGTGCACCTGGTCCCCAAGGCTTCCAAGGTCCCCCTGGTGAGCCTGGCGAGCCTGGAGCTTCAGGTCCCATGGGTCCCCGAGGTCCCCCAGGTCCCCCTGGAAAGAATGGAGATGATGGGGAAGCTGGAAAACCTGGTCGTCCTGGTGAGCGTGGGCCTCCTGGGCCTCAGGGTGCTCGAGGATTGCCCGGAACAGCTGGCCTCCCTGGAATGAAGGGACACAGAGGTTTCAGTGGTTTGGATGGTGCCAAGGGAGATGCTGGTCCTGCTGGTCCTAAGGGTGAGCCTGGCAGCCCTGGTGAAAATGGAGCTCCTGGTCAGATGGGCCCCCGTGGCCTGCCTGGTGAGAGAGGTCGCCCTGGAGCCCCTGGCCCTGCTGGTGCTCGTGGAAATGATGGTGCTACTGGTGCTGCCGGGCCCCCTGGTCCCACCGGCCCCGCTGGTCCTCCTGGCTTCCCTGGTGCTGTTGGTGCTAAGGGTGAAGCTGGTCCCCAAGGGCCCCGAGGCTCTGAAGGTCCCCAGGGTGTGCGTGGTGAGCCTGGCCCCCCTGGCCCTGCTGGTGCTGCTGGCCCTGCTGGAAACCCTGGTGCTGATGGACAGCCTGGTGCTAAAGGTGCCAATGGTGCTCCTGGTATTGCTGGTGCTCCTGGCTTCCCTGGTGCCCGAGGCCCCTCTGGACCCCAGGGCCCCGGCGGCCCTCCTGGTCCCAAGGGTAACAGCGGTGAACCTGGTGCTCCTGGCAGCAAAGGAGACACTGGTGCTAAGGGAGAGCCTGGCCCTGTTGGTGTTCAAGGACCCCCTGGCCCTGCTGGAGAGGAAGGAAAGCGAGGAGCTCGAGGTGAACCCGGACCCACTGGCCTGCCCGGACCCCCTGGCGAGCGTGGTGGACCTGGTAGCCGTGGTTTCCCTGGCGCAGATGGTGTTGCTGGTCCCAAGGGTCCCGCTGGTGAACGTGGTTCTCCTGGCCCTGCTGGCCCCAAAGGATCTCCTGGTGAAGCTGGTCGTCCCGGTGAAGCTGGTCTGCCTGGTGCCAAGGGTCTGACTGGAAGCCCTGGCAGCCCTGGTCCTGATGGCAAAACTGGCCCCCCTGGTCCCGCCGGTCAAGATGGTCGCCCCGGACCCCCAGGCCCACCTGGTGCCCGTGGTCAGGCTGGTGTGATGGGATTCCCTGGACCTAAAGGTGCTGCTGGAGAGCCCGGCAAGGCTGGAGAGCGAGGTGTTCCCGGACCCCCTGGCGCTGTCGGTCCTGCTGGCAAAGATGGAGAGGCTGGAGCTCAGGGACCCCCTGGCCCTGCTGGTCCCGCTGGCGAGAGAGGTGAACAAGGCCCTGCTGGCTCCCCCGGATTCCAGGGTCTCCCTGGTCCTGCTGGTCCTCCAGGTGAAGCAGGCAAACCTGGTGAACAGGGTGTTCCTGGAGACCTTGGCGCCCCTGGCCCCTCTGGAGCAAGAGGCGAGAGAGGTTTCCCTGGCGAGCGTGGTGTGCAAGGTCCCCCTGGTCCTGCTGGTCCCCGAGGGGCCAACGGTGCTCCCGGCAACGATGGTGCTAAGGGTGATGCTGGTGCCCCTGGAGCTCCCGGTAGCCAGGGCGCCCCTGGCCTTCAGGGAATGCCTGGTGAACGTGGTGCAGCTGGTCTTCCAGGGCCTAAGGGTGACAGAGGTGATGCTGGTCCCAAAGGTGCTGATGGCTCTCCTGGCAAAGATGGCGTCCGTGGTCTGACTGGCCCCATTGGTCCTCCTGGCCCTGCTGGTGCCCCTGGTGACAAGGGTGAAAGTGGTCCCAGCGGCCCTGCTGGTCCCACTGGAGCTCGTGGTGCCCCCGGAGACCGTGGTGAGCCTGGTCCCCCCGGCCCTGCTGGCTTTGCTGGCCCCCCTGGTGCTGACGGCCAACCTGGTGCTAAAGGCGAACCTGGTGATGCTGGTGCTAAAGGCGATGCTGGTCCCCCTGGCCCTGCCGGACCCGCTGGACCCCCTGGCCCCATTGGTAATGTTGGTGCTCCTGGAGCCAAAGGTGCTCGCGGCAGCGCTGGTCCCCCTGGTGCTACTGGTTTCCCTGGTGCTGCTGGCCGAGTCGGTCCTCCTGGCCCCTCTGGAAATGCTGGACCCCCTGGCCCTCCTGGTCCTGCTGGCAAAGAAGGCGGCAAAGGTCCCCGTGGTGAGACTGGCCCTGCTGGACGTCCTGGTGAAGTTGGTCCCCCTGGTCCCCCTGGCCCTGCTGGCGAGAAAGGATCCCCTGGTGCTGATGGTCCTGCTGGTGCTCCTGGTACTCCCGGGCCTCAAGGTATTGCTGGACAGCGTGGTGTGGTCGGCCTGCCTGGTCAGAGAGGAGAGAGAGGCTTCCCTGGTCTTCCTGGCCCCTCTGGTGAACCTGGCAAACAAGGTCCCTCTGGAGCAAGTGGTGAACGTGGTCCCCCTGGTCCCATGGGCCCCCCTGGATTGGCTGGACCCCCTGGTGAATCTGGACGTGAGGGGGCTCCTGGTGCCGAAGGTTCCCCTGGACGAGACGGTTCTCCTGGCGCCAAGGGTGACCGTGGTGAGACCGGCCCCGCTGGACCCCCTGGTGCTCCTGGTGCTCCTGGTGCCCCTGGCCCCGTTGGCCCTGCTGGCAAGAGTGGTGATCGTGGTGAGACTGGTCCTGCTGGTCCCGCCGGTCCTGTCGGCCCTGTTGGCGCCCGTGGCCCCGCCGGACCCCAAGGCCCCCGTGGTGACAAGGGTGAGACAGGCGAACAGGGCGACAGAGGCATAAAGGGTCACCGTGGCTTCTCTGGCCTCCAGGGTCCCCCTGGCCCTCCTGGCTCTCCTGGTGAACAAGGTCCCTCTGGAGCCTCTGGTCCTGCTGGTCCCCGAGGTCCCCCTGGCTCTGCTGGTGCTCCTGGCAAAGATGGACTCAACGGTCTCCCTGGCCCCATTGGGCCCCCTGGTCCTCGCGGTCGCACTGGTGATGCTGGTCCTGTTGGTCCCCCCGGCCCTCCTGGACCTCCTGGTCCCCCTGGTCCTCCCAGCGCTGGTTTCGACTTCAGCTTCCTGCCCCAGCCACCTCAAGAGAAGGCTCACGATGGTGGCCGCTACTACCGGGCTGATGATGCCAATGTGGTTCGTGACCGTGACCTCGAGGTGGACACCACCCTCAAGAGCCTGAGCCAGCAGATCGAGAACATCCGGAGCCCAGAGGGCAGCCGCAAGAACCCCGCCCGCACCTGCCGTGACCTCAAGATGTGCCACTCTGACTGGAAGAGTGGAGAGTACTGGATTGACCCCAACCAAGGCTGCAACCTGGATGCCATCAAAGTCTTCTGCAACATGGAGACTGGTGAGACCTGCGTGTACCCCACTCAGCCCAGTGTGGCCCAGAAGAACTGGTACATCAGCAAGAACCCCAAGGACAAGAGGCATGTCTGGTTCGGCGAGAGCATGACCGATGGATTCCAGTTCGAGTATGGCGGCCAGGGCTCCGACCCTGCCGATGTGGCCATCCAGCTGACCTTCCTGCGCCTGATGTCCACCGAGGCCTCCCAGAACATCACCTACCACTGCAAGAACAGCGTGGCCTACATGGACCAGCAGACTGGCAACCTCAAGAAGGCCCTGCTCCTCCAGGGCTCCAACGAGATCGAGATCCGCGCCGAGGGCAACAGCCGCTTCACCTACAGCGTCACTGTCGATGGCTGCACGAGTCACACCGGAGCCTGGGGCAAGACAGTGATTGAATACAAAACCACCAAGACCTCCCGCCTGCCCATCATCGATGTGGCCCCCTTGGACGTTGGTGCCCCAGACCAGGAATTCGGCTTCGACGTTGGCCCTGTCTGCTTCCTGTAA\n"
     ]
    }
   ],
   "source": [
    "GI_id = \"NM_000088.3\"\n",
    "Entrez.email = \"Andrew.JudellHalf001@umb.edu\" # let NCBI know who you are\n",
    "handle = Entrez.efetch(db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=GI_id) # db = database to query, rettype = retrieval type, retmode = retrieval mode \n",
    "seq_record = SeqIO.read(handle, \"gb\") #read the contents \"fetched\" from NCBI\n",
    "print seq_record\n",
    "print type(seq_record)\n",
    "handle.close()\n",
    "x = 0\n",
    "for f in seq_record.features: \n",
    "    x += 1\n",
    "    if f.type == \"CDS\":\n",
    "        cds_start = f.location.start\n",
    "        cds_end = f.location.end\n",
    "        regionplot =ntseq[cds_start:cds_end]\n",
    "# for each feature in this record's features, using the .features attribute of nt_seq_record\n",
    "    # ask if the feature is a CDS, using the .type attribute of the feature\n",
    "        # if so, assign the start location to the variable cds_start, using the .location.start attribute of the feature\n",
    "        # if so, assign the end location to the variable cds_end, using the .location.end attribute of the feature\n",
    "# print the value of cds_start\n",
    "# print the value of cds_end\n",
    "print cds_start\n",
    "print cds_end\n",
    "print regionplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "collapsed": false,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZAAAAEoCAYAAAB2JqfUAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XmYHFW9xvHvm4QtLGEgJBFCgoCyiSwCsnhhBEFxAa4L\nIqIsigsIiCIkbkR52NQrKldcAZMIKqBsipcQw7CIiBJ2QgRZDSFsgQARBPK7f5zTSU3TPdOpTKd7\nZt7P8/STqlOnqk6f6dSv6lTVOYoIzMzMltaQVhfAzMz6JwcQMzMrxQHEzMxKcQAxM7NSHEDMzKwU\nBxAzMyvFAcRsOZJ0rqSnJd3Y6rKYLSsHkAFM0gGSbpT0vKTHJP1F0mer8uwg6Q+S5kt6Muc/pEVF\n7hOSHpC0e6vLUU3S24A9gHUjYscayw+WtEjScVXpj0jaNU9PynmOqspzTE7/egPl2EPSrPy7+JOk\ncVXLT8+/hScknVa1bLykGZJekHS3pD2qlh8o6UFJz0n6naQ1C8s6JP0mb/txSVMlrVZYvrukmyU9\nK+k+SYf38B1Ol/RwzvuApAm9fW/rew4gA5SkLwJnAKcDoyNiDPAZYGdJK+Q8OwF/Aq4GNoqIkcBn\ngXcuh/INbfY+2tAGwIMR8WIPeZ4Gjpe0ap3lAcwGPl6V/vGc3iNJawO/Bb4CrAXcDPymsPzTwD7A\nlsCbgfdJ+lRhE7/K66wFfBW4KG8TSVsAPwY+CowG/g38qLDuycAIYDywETAGmJTXHQb8DvhRRIwA\nDgC+K2nLOl/lbGDznHdn4CBJ+/X2/a2PRYQ/A+wDrAE8D+zXS77rgB8sxXYPBq4HzgSeAe4Gdq/a\n78+BR4FHgJMAVa37XeBJ4Js5/fC8nQXAncDWOf11wEXA48A/gaMK+zmRdNCbnNe7A9g2L5sCvAq8\nkJcdl9MvAOYC84Eu0sGnsr21gMuBZ4G/5nJfV1i+KTANeAqYBXyohzp6HXBpzvsP4JM5/TDSAfXl\nXK4T69TvdXn9rxfSHwF2LXz3qcBdwGY5bfM8P6W4Xp3yHQ5cX5gfDiwE3pjn/1wpc54/FLghT78x\nf4dVC8uvAT6Vp08GfllYtiHwUiU/cAXwmcLyI4A/5ulR+e+2cmH5TcCHG/hdrgfcXvlb+7P8Pr4C\nGZh2AlYELquXQdIqOd9vl3LbbwXuBdYmnT0WmykmA/8hHTi2AfYEPlm17n2kg8XJkj4EfB04KCLW\nIJ35PiVJpAP6LaQD8h7AMZL2LGzrfcD5pDPay4EfAkTEx4GHgfdGxBoR8Z2c/wrSWe8oYCZwXmFb\nZwHP5WWHkA7kketpOCl4/BIYSToz/qGkTevUz2/y/scAHwJOkdQZEeeQrgD/ksv1jTrrB/A14PPF\n5p8aeabmcpL/nQKoTv6iLYDbFm8oYiHpb7JFreV5urJsc+D+iHihzvLqbd9PCiBvzEk/JF3RrCmp\nA/gA6e9CRDxOuro5TNKQfHU8jnTSUZOkEyQ9Rwqww0m/B1uOHEAGppHAkxGxqJIg6c/5PsfC3Bbf\nQfr7z13Kbc+LiB9ExKsRcQGp2eQ9kkYBewPHRsSLEfEk8D3gI4V150TEWRGxKCJeAj4BfCsiZkI6\n4ETEI8D2wMiIODnv50HSlc0BhW1dHxFXRjoFnUpqbinqdjCNiF9ExMKIeBn4JrCVpNUlDQHeTzpz\nfykiZpECYcV7gQciYkokt5GaWj5UXTGSxpKC8gkR8XLO+3Ne29zUo4i4HbgKOKGHbOcBB+SmnwNI\nAa4Rq5GutIoWAKvXWb4gp5VZt3r5TNKJzVPAE8ArdG/i+jXphOIl0pXNVyJiTr0vEhGnR8TqpJOV\nqTX2bU3mADIwPQWMzAdHACJil4joyMuGkJpyFpHO8JdG9X/oh4B1Se3aKwBz81NG80nt4SMLeR+p\nWnd9UvNUtfHAenk7lW1NJF0hVDxWmF4IrFz8vkX5jPa0fGP2GeAB0ln8SGAdYCjwrzrlHA/sWFWW\nA0lXGNXWBZ7OZ/UVD5GaWJbW14HP5sD8GjnQ/hM4BfhHTwfaKs+TmhqLRpCuwGotH5HTyqxbvfxC\n0gnHqjnf/eQrwXxF9xvS1egKpKuZEyTt3dsXyoH6RdKJgS1HDiAD019IZ3H71ssQEf/O+T6wlNuu\nPhiOY8k9jxeBtSNirYjoiIg1I6J4ZVDd9fMjpGalao+QmkrWKmxrRES8r8EyVu/nQFKT1+4RsSbp\nZrbyp3ImPLaQf/2qsnRVlWWNiDiyxn4fBdaqugE+jtcG3d6/QMRs0pXOV2p8n4opwBfofsXUm7uA\nrSszuawbke4/VZZvVci/dU6rLNuw6vttVbV88bqSNiKdVMwu5P1JvkJdSDrBqASILYB7ImI6QETc\nC/yhsLw3w0hNp7YcOYAMQBHxLOls7CxJH5C0mpKtSW3FFccDh0j6oqS1ACRtJelXPWx+lKSjJA3L\n9zA2Ba6IiMdI9wrOyE1DkrRh5fHTOn4OHCdp27zvjSStT7p5+pyk4yWtLGmopC0kbdfDtopNVo/R\n/WCyOimgzs8Hv1PJB+XczPc7YJKkVfKZcLHJ6ffAGyUdlL/zCpK2q3UPJCL+BdwAnCppJUlvJjXT\nTe2h3D35Jukmdr17Ib8B9iKd2TfqYmALSf8taSXSTflb8wEbclCStK6k9UgB6lxYfFC/FTgxf7/3\nA29iyX2080j3OHbJ9fxN4LeFK7KbgE/mv+kqwKdJN78h3e/aWNLbYXHweS/d78eQl0nSpyr3iCTt\nABwJTF+KerC+0Oq7+P4070O6//BXUtPCPNIVxyeAYYU825FuZM4nPR31F1IzQq3tVZ4S+gHpKax7\ngD0Ky1cn3ZB+JG/vZmD/wrrX1tjmp/J2FpAOJlvl9DGkm6JzSc1uN5Cf+CId9KYUtjGe9ATPkDy/\nD6np6GnSAXA4cEnexwPAQTn/hjn/SFKgeCbX16nAVYXtvyEvf5x0xTIdeHOdOlqXdFP/KdLDBodX\n1d9r6qCn5aQbz6/S/SmsKXXW7/UprJxvd9LTZC8AM4BxVctPy+V/Eji1atk40mPfC/M23l61/IBc\n98+RAvOaVX+ny/J2n8y/u40Kyz9IeqLuWdKDCKcUlh0I3JGnBfwxb2NB/v2c0Or/b4PxU3nE0qxX\nkg4GPhERPV1V9HtKL8+NjohDW10Ws3bmJiwb9CRtUnlhLTeHfIJ09mxmPXAAMUtNb7+T9DzpXYRv\nR8TlLS5TaZImKnUlsqDq84dWl80GFjdhmZlZKb4CMTOzUhxAbKkp9fq6TM/cSzpRUtnHW5d2X7tJ\nqn6J0ZosN6Nt0OpyWPM4gFgZfdXu2dB2lMbQWNa3jNuurTYHtlfz/YnKPYuPFZavKOkcpS7LH5V0\n7DLsqy/qcKlExOqRuqGxAWpYqwtg/VIjnfZZgaShEfFqjUVzImJcjXSAb5DeEl+f9H7J1ZLuiohp\nzSpnRQ/lNVvMVyCDWH67e5qkpyTNVR6UR9L2km5Q6nxxjqQzlTrtq7WNlSX9j9IgQvMlXZvfUn5N\ns5F6GOhJ0gW5DPMldUnaLKcfThpf4vh8hn5pTn+dpIuUBib6pwoDLOUy/UKp76o7SZ0z9lQPO0u6\nKe/7r0o9wSJpf0l/q8p7rKRL8vSKkr4j6aFc9rPy292Lm82U3qafC5zTUxnq+Dip2/sFEXEP8FNS\nb8G1vkNlf1+QNC//3Q7pqQ5rbGORpCMk/YPUFT2SNi38RmYp9T5Qyb+WpMvzFdJfJZ0k6bqq7W2Y\np9eQNCX/vR6Q9JVCvoMlXSfp2/lv9k9J7yosPySnLcj/FjvotFZq9ZuM/rTmQ+o59VHg86QeUlcF\nts/LtgV2IF1pjCP1cXR0Yd1FLHmL+4ekt5nH5Pw7kvo/2g14uGqfD1D/bfJDSG+Mr0AaM+SWwrJz\nyeOH5HkBfyf1EzWU1LfVfcCeeflppN5cR5D67rqjuiyFbXWQ3lg/kHRCdUCe7wBWIb0VXXxb+iby\neCCkAbsuyftZlTSOx8l52W6ksT9Oyd9ppRr73o3Uf9hcUseI3wWG52Vr5npep5D//cBtdb5HZX8n\n5jrZm/Sm+YhadVhnG4uAK/O+V8p/j4dJgUykvqyeADbN+X9N6i1gJWCznPfawvaKb/tPIXWjMpz0\nRvps4NC87GBSVzOH5f18hnRlRs7/LLBxnh9NHgfFn9Z/Wl4Af1r0h08HypsbzHsMqU+jyvwiUl9T\nInVp8aYa6yxVAKnKVzl4rp7nqwPIDqSR/YrrTADOztP/JAeTPH94dVkKyw4CbqxKuwH4eJ6eCnw1\nT78hH8xWyvPPA68vrLcTqRPIYnBYoYd6HVU4GI8nBb0f5fmx+QC8YiH/Oyrbr1PfL5C7c8lp84Ad\natVhnW0sAnYrzO8PXFOV58ek8UqGkMZ+2biw7KSqAFL5nQzJAWKTwrJPATPy9MGkHoUry1bJ647K\nAeRp4L8pDDblT3t83IQ1eNXrSh1Jb8hNE3OVuj8/me7dsleMJJ193r8sBVHP3a3XMp6eu3tfl+7d\nsz/Uw+7XrbG82AX7+SwZ0+RA4JKIeEnSOqSD282VcpD6Z1q7sJ0nIo0/UlNEPB6paYqIeIjUuWWl\nd+RKF+rVXas/R31PRWEMGFJwX61e5jqK9Tae2l3ZjyZ1gz+M+t3gF43MeR8upFV3c7+4e/5IPUUD\nrBapI8YPk4Zanpt/l5ss5XeyJnEAGbzqdaUOaZCfWaSmmzVJTUW1bpw/STrLrrWdFyj0/Ks0Bvo6\ndfb3Uep3tw61u4Hvqbv3R+neJfv4Ovut5N2gKq3YBftVwDqStiJdtVVGvXuSdIDeolCONSON0V1R\n5smvIQAR8QypaavYtXqx6/Sl1WhZivnqdWX/OVJT1svU7wa/6Mmct/h3GE+D3dxHxFURsRepmXQ2\n8LPGvoo1mwPI4PV7YIyko/PN4NWU+oGC1LXHgohYqNRt+WdrbSAigtQ08t18U3uIpB0lrUC6Cbuy\npL3zDfivku611LIadbpbz+bRvXv23rp7vxCYqDR06ljgcz3UwxXAGyQdkLfzYVJ7/u/zd3wlb+/b\npPsiVxW++8+A7+WrESStJ2mvHvbVjaROSePy9PqkezeXFLJMBb6av8dmpKa4cxvdfpXqOmxEva7s\nN4neu8FfLOe9gDSM8WqSxgPH0kA395JGSdpHaWjhl0lXZn46rE04gAxSEfE8aczyfUjNB/8AOvPi\n44CPSloA/IR0s7Tb6oXp40g3qf9G6gL8NFI7/ALgCOBsUjPHc3Rv7iiaQmremEMa2OiGquVnk8aw\neFrS7/IB6b2kwY4eIHWz/jOWNPd8I2/vAeD/8vbr1cPTeVvHkc6UjwPek9MrfkUal/2CqiaiE0g3\n72/MTW/TWDL+dyO2AW5Q6oPretJYG8cUlp9Iah58iPSgwmkRcdVSbL/4d+pWhw3kr/xG9iJdeT2a\nP6eRmi0BjiLdr5pLGtTqfNKJQK3tHU26YrsfuBb4ZUT0FAwr6w4hdck/h/T32ZU6JzS2/DW1LyxJ\nZ5P+c86LPDKdpG+RmiteIrXBH5oPNkiaSHoS4xXgmFgOz7ubWd+Qu8EfdJp9BXIu8M6qtGmkduOt\nSQPuTASQtDnpqY/NSI8gniXJL6yZtSm5G/xBr6kBJCKuJ41MV0ybXmgGuJElN+H2AX4dEa9E6v7g\nXtLjmmbWngZUN/i29FrdlclhpB8epEf6/lJYNofuj/mZWRuJiL+T3o2xQaplN9FzVwYvR8Sves1s\nZmZtpyVXILmPnncDxX6R5tD9OfKx1HlOXFLb9axqZtYfRESf3VteHgGk+EIYuZO0LwG7RkTxkb/L\ngPMknUFqutqY9Lx/Tb09PTZ//nyGDx/eY56B4KSTTuJrX/taq4vRFlwXS7gulminuli4cCEdHR0t\n239fP5fU1AAi6XzSuwVrS3qY9Fz7l0kvlF2Vv8yNEXFERNwt6QLgbtILQ0dEM58xNjOzZdLUABIR\nB9ZIrvvyUEScSnoL2czM2pzfRO/ndt1111YXoW24LpZwXSzhumiepr6J3iySem3dGiz3QMys/2iH\neyB9eRPdVyBmZlaKA4iZmZXiAGJmZqU4gJiZWSkOIGZmVooDiJmZleIAYmZmpTiAmJlZKQ4gZmZW\nigOImZmV4gBiZmalOICYmVkpDiBmZlaKA4iZmZXiAGJmZqU4gJiZWSkOIGZmVooDiJmZleIAYmZm\npTiAmJlZKQ4gZmZWigOImZmV4gBiZmalOICYmVkpDiBmZlaKA4iZmZXS1AAi6WxJ8yTdXkjrkDRN\n0mxJV0oaUVg2UdK9kmZJ2quZZTMzs2XT7CuQc4F3VqVNAKZHxCbADGAigKTNgf2BzYC9gbMkqcnl\nMzOzkpoaQCLiemB+VfK+wOQ8PRnYL0/vA/w6Il6JiAeBe4Edmlk+MzMrrxX3QEZFxDyAiHgMGJXT\n1wMeKeSbk9PMzKwNtcNN9Gh1AczMbOkNa8E+50kaHRHzJI0BHs/pc4D1C/nG5rSaJk2atHi6s7OT\nzs7Ovi+pmVk/1tXVRVdXV9O2r4jmXgBI2gC4PCK2zPOnA09HxOmSTgA6ImJCvol+HvBWUtPVVcAb\nokYBJdVK7mb+/PkMHz68T7+LmdmyWLhwIR0dHS3bvyQios8eTmrqFYik84FOYG1JDwMnAqcBF0o6\nDHiI9OQVEXG3pAuAu4GXgSN6jRJmZtYyTb8CaQZfgZhZfzTQrkDa4Sa6mZn1Qw4gZmZWigOImZmV\n4gBiZmalOICYmVkpDiBmZlaKA4iZmZXiAGJmZqU4gJiZWSkOIGZmVooDiJmZleIAYmZmpTiAmJlZ\nKQ4gZmZWigOImZmV4gBiZmalOICYmVkpDiBmZlaKA4iZmZXiAGJmZqU4gJiZWSkOIGZmVooDiJmZ\nleIAYmZmpTiAmJlZKQ4gZmZWigOImZmV4gBiZmalOICYmVkpLQsgko6VdKek2yWdJ2lFSR2Spkma\nLelKSSNaVT4zM+tZSwKIpHWBo4BtI+LNwDDgI8AEYHpEbALMACa2onxmZta7VjZhDQVWlTQMWAWY\nA+wLTM7LJwP7tahsZmbWi5YEkIh4FPgf4GFS4Hg2IqYDoyNiXs7zGDCqFeUzM7PeDWvFTiWtSbra\nGA88C1wo6aNAVGWtnl9s0qRJi6c7Ozvp7Ozs83KamfVnXV1ddHV1NW37iqh7jG7eTqUPAu+MiMPz\n/MeAHYHdgc6ImCdpDHB1RGxWY/3ordzz589n+PDhfV94M7OSFi5cSEdHR8v2L4mIUF9tr1X3QB4G\ndpS0siQBewB3A5cBh+Q8BwOXtqZ4ZmbWm5Y0YUXETZIuAm4BXs7//hRYHbhA0mHAQ8D+rSifmZn1\nriVNWMvKTVhm1h+5CcvMzAwHEDMzK8kBxMzMSnEAMTOzUhxAzMysFAcQMzMrxQHEzMxKcQAxM7NS\nHEDMzKwUBxAzMyvFAcTMzEpxADEzs1IcQMzMrBQHEDMzK8UBxMzMSnEAMTOzUhoKIJL+1EiamZkN\nHj0OaStpZWA4MFJSB1AZyWoNYL0ml83MzNpYb2Oifxr4PLAucDNLAsgC4H+bWC4zM2tzDY2JLumo\niDhzOZSnIR4T3cz6o4E2JnpvVyAARMSZknYGNiiuExFT+qogZmbWvzQUQCRNBTYCbgVezckBOICY\nmQ1SDQUQYDtg817bjczMbNBo9D2QO4ExzSyImZn1L41egYwE7pZ0E/BSJTEi9mlKqczMrO01GkAm\nNbMQZmbW/zT6FNY1ksYDb4iI6ZKGA0ObWzQzM2tnjXZlcjhwEfCTnLQecEmzCmVmZu2v0ZvoRwK7\nkN5AJyLuBUYty44ljZB0oaRZku6S9FZJHZKmSZot6UpJI5ZlH2Zm1jyNBpCXIuI/lRlJw0jvgSyL\n7wNXRMRmwFbAPcAEYHpEbALMACYu4z7MzKxJGg0g10j6MrCKpD2BC4HLy+5U0hrAf0XEuQAR8UpE\nPAvsC0zO2SYD+5Xdh5mZNVejAWQC8ARwB6mDxSuAry7Dfl8PPCnpXEkzJf0035gfHRHzACLiMZax\nmczMzJqn0cd4VwHOiYifAUgamtMWLsN+twWOjIi/SzqDFKSqm8XqNpNNmjRp8XRnZyednZ0li2Jm\nNjB1dXXR1dXVtO032hvvjcA7IuL5PL8aMC0idi61U2k08JeI2DDPv40UQDYCOiNinqQxwNX5Hkn1\n+u6N18z6nYHWG2+jTVgrV4IHQJ4ufXTOzVSPSHpjTtoDuAu4DDgkpx0MXFp2H2Zm1lyNNmG9IGnb\niJgJIOktwL+Xcd9HA+dJWgG4HziU9HLiBZIOAx4C9l/GfZiZWZM0GkCOAS6U9ChpVMIxwIeXZccR\ncRuwfY1F71iW7ZqZ2fLRawCRNARYEdgU2CQnz46Il5tZMDMza2+9BpCIWCTphxGxDalbdzMzs4Zv\nov9J0gck9dndezMz698afYz3OWBV0nC2/ybdB4mIWKO5xatbHj/Ga2b9zkB7jLfR7txX76sdmpnZ\nwNBod+6SdJCkr+X59SXt0NyimZlZO2v0HshZwE7AgXn+eeCHTSmRmZn1C42+B/LWiNhW0i0AETFf\n0opNLJeZmbW5Rq9AXs4dKAaApHWARU0rlZmZtb1GA8gPgIuBUZJOBq4HTmlaqczMrO01+hTWeZJu\nJnV6KGC/iJjV1JKZmVlb6zGASFoZ+AywMWkwqZ9ExCvLo2BmZtbeemvCmgxsRwoeewPfaXqJzMys\nX+itCWvziNgSQNLZwE3NL5KZmfUHvQWQxT3uRsQrbdUVVi9laV1nAWZmta3U6gL0sR77wpL0KvBC\nZZYl46C7Lywzs6U0qPrCioihfbUjMzMbWBp9D8TMzKwbBxAzMyvFAcTMzEpxADEzs1IcQMzMrBQH\nEDMzK8UBxMzMSnEAMTOzUhxAzMysFAcQMzMrxQHEzMxKaWkAkTRE0kxJl+X5DknTJM2WdKWkEa0s\nn5mZ1dfqK5BjgLsL8xOA6RGxCTADmNiSUpmZWa9aFkAkjQXeDfy8kLwvaRRE8r/7Le9ymZlZY1p5\nBXIG8CWgOLDH6IiYBxARjwGjWlEwMzPrXW8jEjaFpPcA8yLiVkmdPWStO2rUpEmTFk93dnbS2dnT\nZszMBp+uri66urqatv0eRyRs2k6lU4CDgFdIoxyuDlwMbAd0RsQ8SWOAqyNisxrre0RCM+t3BtqI\nhC1pwoqIL0fEuIjYEDgAmBERHwMuBw7J2Q4GLm1F+czMrHetfgqr2mnAnpJmA3vkeTMza0MtacJa\nVm7CMrP+yE1YZmZmOICYmVlJDiBmZlaKA4iZmZXiAGJmZqU4gJiZWSkOIGZmVooDiJmZleIAYmZm\npTiAmJlZKQ4gZmZWigOImZmV4gBiZmalOICYmVkpDiBmZlaKA4iZmZXiAGJmZqU4gJiZWSkOIGZm\nVooDiJmZleIAYmZmpTiAmJlZKQ4gZmZWigOImZmV4gBiZmalOICYmVkpDiBmZlaKA4iZmZXSkgAi\naaykGZLuknSHpKNzeoekaZJmS7pS0ohWlM/MzHrXqiuQV4AvRMQWwE7AkZI2BSYA0yNiE2AGMLFF\n5TMzs160JIBExGMRcWuefh6YBYwF9gUm52yTgf1aUT4zM+tdy++BSNoA2Bq4ERgdEfMgBRlgVOtK\nZmZmPWlpAJG0GnARcEy+EomqLNXzZmbWJoa1aseShpGCx9SIuDQnz5M0OiLmSRoDPF5v/UmTJi2e\n7uzspLOzs4mlNTPrf7q6uujq6mra9hXRmpN8SVOAJyPiC4W004GnI+J0SScAHRExoca60Vu558+f\nz/Dhw/u62GZmpS1cuJCOjo6W7V8SEaE+214rAoikXYBrgTtIzVQBfBm4CbgAWB94CNg/Ip6psb4D\niJn1OwMtgLSkCSsi/gwMrbP4HcuzLGZmVk7Ln8IyM7P+yQHEzMxKcQAxM7NSHEDMzKwUBxAzMyvF\nAcTMzEpxADEzs1IcQMzMrBQHEDMzK8UBxMzMSnEAMTOzUhxAzMysFAcQMzMrxQHEzMxKcQAxM7NS\nHEDMzKwUBxAzMyvFAcTMzEpxADEzs1IcQMzMrBQHEDMzK8UBxMzMSnEAMTOzUhxAzMysFAcQMzMr\nxQHEzMxKcQAxM7NSHEDMzKyUtgwgkt4l6R5J/5B0QqvLY2Zmr9V2AUTSEOB/gXcCWwAfkbRpa0vV\nvq655ppWF6FtuC6WcF0s4bponrYLIMAOwL0R8VBEvAz8Gti3xWVqW9dee22ri9A2XBdLuC6WcF00\nTzsGkPWARwrz/8ppZmbWRtoxgJiZWT+giGh1GbqRtCMwKSLelecnABERpxfytFehzcz6iYhQX22r\nHQPIUGA2sAcwF7gJ+EhEzGppwczMrJthrS5AtYh4VdLngGmkJrazHTzMzNpP212BmJlZ/9DvbqIP\nhpcMJZ0taZ6k2wtpHZKmSZot6UpJIwrLJkq6V9IsSXsV0reVdHuuq+8t7++xrCSNlTRD0l2S7pB0\ndE4fjHWxkqS/Srol18WJOX3Q1UWFpCGSZkq6LM8PyrqQ9KCk2/Jv46actnzqIiL6zYcU8O4DxgMr\nALcCm7a6XE34nm8DtgZuL6SdDhyfp08ATsvTmwO3kJojN8j1U7my/CuwfZ6+Anhnq7/bUtbDGGDr\nPL0a6d7YpoOxLnK5h+d/hwI3kt6ZGpR1kct+LPBL4LI8PyjrArgf6KhKWy510d+uQAbFS4YRcT0w\nvyp5X2Bynp4M7Jen9wF+HRGvRMSDwL3ADpLGAKtHxN9yvimFdfqFiHgsIm7N088Ds4CxDMK6AIiI\nhXlyJdIBIBikdSFpLPBu4OeF5EFZF4B4bWvScqmL/hZABvNLhqMiYh6kAyswKqdX18mcnLYeqX4q\n+nVdSdqAdFV2IzB6MNZFbrK5BXgMuCr/Zx+UdQGcAXyJFEQrBmtdBHCVpL9J+mROWy510XZPYVnD\nBs3TD5JWAy4CjomI52u8BzQo6iIiFgHbSFoDuFjSFrz2uw/4upD0HmBeRNwqqbOHrAO+LrJdImKu\npHWAaZJms5x+F/3tCmQOMK4wPzanDQbzJI0GyJebj+f0OcD6hXyVOqmX3q9IGkYKHlMj4tKcPCjr\noiIiFgACqSV1AAAFEklEQVRdwLsYnHWxC7CPpPuBXwG7S5oKPDYI64KImJv/fQK4hNTUv1x+F/0t\ngPwN2FjSeEkrAgcAl7W4TM2i/Km4DDgkTx8MXFpIP0DSipJeD2wM3JQvW5+VtIMkAR8vrNOfnAPc\nHRHfL6QNurqQNLLyJI2kVYA9SfeEBl1dRMSXI2JcRGxIOgbMiIiPAZczyOpC0vB8hY6kVYG9gDtY\nXr+LVj9BUOKJg3eRnsa5F5jQ6vI06TueDzwKvAQ8DBwKdADT83efBqxZyD+R9DTFLGCvQvpb8o/p\nXuD7rf5eJephF+BV0tN2twAz899/rUFYF1vm738rcDvwlZw+6Oqiql52Y8lTWIOuLoDXF/5/3FE5\nJi6vuvCLhGZmVkp/a8IyM7M24QBiZmalOICYmVkpDiBmZlaKA4iZmZXiAGJmZqU4gFi/J+ng/LZt\ns7b/OkkX5OmtJO1dWPY+Scf30X5WltSVX+RC0qu5u/JbJF1SyHe1pHE11t8+56189isseyD/u4Kk\nayT5/74tM/eFZQPBIcCdpE4G+1ykriL2z7NbA9sBf8zLLie9Ad0XDgN+G0teznohIrZdivXvAN4S\nEYtyQL1N0mWR+tCKXN6XJU0nvcF9fh+V2wYpn4VYv5C7r7lb0k8l3Snp/5QGWfoA6YD+y3y2vlLV\neldL+l4+I79d0vY5vUPSxXkgnhskbZnTd8t5Z0q6WdKqed935H65vgnsn5d/KF/9nFko458k3Srp\nqtzlOJLOlfR9SX+WdJ+k99f5mh+le/cRqpPvKdIb+t1ExIs5WACsAiwqLH6iMH1p3pfZMnEAsf5k\nY+DMiHgT8CzwgYj4LfB34MCI2DYiXqqx3ioRsQ1wJKlvLYBvADMjYivgK6TxDwC+CByRz/z/C/h3\nTo+IeAX4OvCbvK8LK8vyv2cC50bE1qSz+zMLZRgTEbsA7yMN9tONpBWA10fEw4XklST9PQe4xePe\nRMQHI6JmR3e5L6M7gduAz1QCSkS8tZDtTmD7WuubLQ0HEOtPHoiIO/L0zaQR1Srqna1D6rGViLgO\nWD13Svg2YGpOvxpYK3dK92fgDElHkUZ5W1Rnm7XsVNlX3vYuhWWX5H3NYsnYDEUjgWeq0sZHxHak\nq4Xv5c7vehQRN+UAuz3w5dzpaHWeRcBLufM9s9IcQKw/KV5dvErj9/CqO3yrFRQEEBGnA58gNQH9\nWdIbl6J8PXUsVyx7rWD3b2Dlbhtb0k33A6Tu27fpVmBpv0Jz27ZV684GngfeVKc8KwEv9lBes145\ngFh/Uu8q4zlgjR7W+zCApLcBz0bEc8B1wEE5vRN4ItJgVRtGxF0R8S3S8AGbVu27p33dAHwkTx+U\n99HQ94iIZ4ChlSsGSWsWpkcCOwN3V61zSURsk5vTZkraQNLQvM54YBPgwdfsXFoLeDIiXnMfxWxp\n+Cks60/qneH/AvixpIXATjXug7woaSbp935oTpsEnCPpNuAF0vgHAJ+X9HbSFc5dpKet1i3s+2pg\nQt7eqVX7ORo4V9JxpJvWlX01OjrcNFLT2gxgM+Ankl4lneidGhH31Fmv4m25bP8hXWV9NiKerpHv\n7cAfetmWWa/cnbsNaJKuBr4YETNbXZbeSNoG+HxEHNzk/fwWOCEi7mvmfmzgcxOWDXT95gwpIm4B\nrq68SNgM+Wmvix08rC/4CsTMzErxFYiZmZXiAGJmZqU4gJiZWSkOIGZmVooDiJmZleIAYmZmpfw/\nlDz0YoYNnHgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f0b5881da10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt # import a set of code written for generating graphics and tell the computer how to display graphs\n",
    "#%pylab inline\n",
    "\n",
    "plt.plot(Y) # plot the values of region\n",
    "plt.axhline(0, color='black') # forces the y-axis to start at 0\n",
    "plt.axhline(50, color='red') # add a red line representing the average GC content of the entire sequence\n",
    "\n",
    "plt.axvspan(cds_start, cds_end, facecolor='grey', alpha=0.05) # create a grey region corresponding to the coding region\n",
    "plt.ylabel(\"Percent\") # label the y-axis\n",
    "plt.xlabel(\"nt position (5'-3')\") # label the x-axis\n",
    "plt.title(\"GC percentage of \" + ntid + \"\\n calculated over 50 nt regions\") # add a title\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Q5: Are there noticeable differences in the GC content between the different regions of the mRNA sequence? Explain."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline \n",
    "import numpy as np #linear algebra/vector/matrix tool\n",
    "import matplotlib.pyplot as plt #plot tool\n",
    "\n",
    "x = np.random.randn(100,1) #length 100 normal random variable vector\n",
    "y = x**2 #square each element of x\n",
    "plt.scatter(x,y)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's all folks!  \n",
    "Please save your notebook and upload the notebook to Blackboard."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
